---
title: "A4-vat-tconv-treg"
author: "JRD"
date: "February 9, 2016"
output: html_document
---

last modified: `r Sys.Date()`
<br>    

#### README
+ this document contains code and explanation for an analysis comparing VAT Tconv, VAT Treg and splenic Tregs
+ for details on how the input files used in this analysis were generated, see *ATAC_mapping.Rmd* and *ATAC_peaks.Rmd*
<br>    


#### R Markdown Specifications
Set global options
```{r setoptions, echo=F}
library(knitr)
opts_chunk$set(echo=F, eval=F, warning=F, message=F)
```
<br>    


Github
https://github.com/jdis24/ATAC.git
<br>    

#### Pseudo
+ [these steps were done in "ATAC_peaks.Rmd"]
+ for each sample, call peaks in each replicate
+ then merge into single peakset
+ for each replicate, count reads in each peak, normalized to total library size
+ export .csv files of peaks and counts

+ [these steps done in this document]
+ plot distributions of read counts 
+ filter for strong peaks (start with >=3 rpm)
+ get genomic annotations of strong peaks 
+ use DiffBind to find differential peaks between VAT Treg and Tconv
+ use merged peaks as input to DiffBind, with .bam files for each replicate
+ export all peaks in the contrast with p-values and/or FDR
+ plot p-value/FDR distributions
+ set cutoffs
+ export differential and similar peaks, with counts in rpm
<br>    


#### Import peak files into peak list
```{r peak_list, eval=T}
# create path to .csv files  
path <- "/Volumes/JD_WD2/JRD_CBDMLab/1_JRD_CBDM_PROJECTS/ProjectFolders/T-TregEpigenome/Data/1_Experimental/ATAC/Master_peaks/R/peaks_with_counts/working/"

# get filenames
filenames1 <- list.files(path = path, pattern = ".csv")
filenames <- paste(path, filenames1, sep="") # update filenames with path

# combine peak table .csv dfs into a list
colClass <- c(rep("factor",2), rep("numeric",3), "character", rep("numeric",4)) # set column classes
pl <- lapply(filenames, read.csv, colClasses = colClass) # [p]eak [l]ist

# name the list elements to match the files
names(pl) <- gsub("\\.csv", "", filenames1)

# convert each df to bed format
pl_bed <- lapply(pl, function(x) {
  x <- x[,c("seqnames","start","end")]
})

# export each df as tdt bed file
lapply(names(pl_bed),
       function(x) write.table(pl_bed[x], file=paste(paste0(path,x),"bed",sep="."), sep = "\t", 
                              quote=F, row.names=F,col.names=c("chr","start","end")))

```
<br>    


#### Generate table of peak numbers for range of rpm threshold values
```{r threshold_table, eval=T, results="asis"}
library(xtable)

# function to compute the number of peaks >=n rpm in each sample
# ARGS: df = dataframe of peak height rpms, peak_height_var = string of var name storing rpm values, cutoffs = numeric vector of rpm cutoffs 
# USAGE: number_peaks_above_nrpm(df, "peak_height_var", cutoffs)
number_peaks_above_nrpm <- function(df, rpm_var, cutoffs) { 
  # create indexing vector representing number of elements in cutoffs
  n <- length(cutoffs) 
  # create output vector whose length = length of cutoffs
  o <- numeric(n) 
  # create variable to store vector of extracted rpm values
  rpm <- df[[rpm_var]]
  # loop over each element of cutoffs and count number rows where peak height >= that element
  for(i in 1:n) { 
    o[i] <- as.numeric(sum(rpm >= cutoffs[[i]])) # calculate the total number of rows where rpm >= ith value in cutoffs  
  }
  # print output vector
  return(o) 
}


# create cutoff vector of rpm values
cutoffs <- c(seq(0,5,by=0.5))

# apply cutoffs to list
peaks_pass_filter <- sapply(pl, number_peaks_above_nrpm, "rpm", cutoffs)

# transform list into dataframe (if doesn't work may have to call base as.data.frame bc of Biocgenerics masking?)
ppf <- as.data.frame(peaks_pass_filter) #[p]eaks [p]ass [f]ilter
ppf <- cbind(ppf,"rpm_cutoff" = cutoffs) # add column for each cutoff

# write to file
write.csv(ppf, file = paste0(path,"peak_totals_rpm_filtered.csv"), row.names=F)

# generate xtable
print(xtable(ppf), type="html", include.rownames=F)
```
<br>    


#### Plot distribution of read counts for each sample
```{r plots, eval=T}
library(reshape2)
library(ggplot2)
library(dplyr)
# set path for plots 
path <- "/Volumes/JD_WD2/JRD_CBDMLab/1_JRD_CBDM_PROJECTS/ProjectFolders/T-TregEpigenome/Data/1_Experimental/ATAC/Analyses/A4-vat-tconv-treg/"

# get peak list into shape for plotting
plm <- melt(pl, id.vars=names(pl[[1]])) # [p]eak [l]ist [m]elted
plm$L1 <- NULL # remove extraneous column

# calculcate log2 rpm values and append as new column
plm$log2_rpm <- log(plm$rpm, 2)

# show number of peaks where rpm =0 
rpm0pks <- plm$log2_rpm == "-Inf"
plm_rpm0pks <- plm[rpm0pks,]

print("Peaks where rpm =0 ...") 
xtable(plm_rpm0pks)

# filter out peaks where rpm = 0
plmf <- plm[!rpm0pks,]  #plm[f]iltered

# draw boxplot of all peaks in each sample
g <- ggplot(data = plmf, aes(x=Sample, y=log2_rpm, fill=Sample))
g + geom_boxplot(varwidth =T) +
  theme_bw() + 
  ylim(-2,10) +
  theme(axis.title.x = element_blank())
ggsave(paste0(path, "a4-boxplot.pdf"), width=5, height=4) # export pdf

# draw boxplot of peaks passing n rpm
# define function to filter peaks and draw plot
bp_nrpm <- function(df, n) {
  df %>%
    filter(rpm >= n) %>% # filter by rpm cutoff 
    ggplot(aes(x=Sample, y=log2_rpm, fill=Sample)) + # plot
    geom_boxplot(varwidth=T) +
    ylim(-2,10) +
    theme_bw() + 
    theme(axis.title.x = element_blank())
}

n <- 3 # set rpm cutoff
bp_nrpm(plmf, n) # draw plot
ggsave(paste0(path,"a4-boxplot-",n,"rpm-filtered.pdf"), width =5, height =4) # export pdf
# ggsave(paste0(path, "a4-boxplot-rpm-filtered.pdf"), width=5, height=4)


# draw histograms of all peaks in each sample
g <- ggplot(data = plmf, aes(x=log2_rpm))
g + geom_histogram(fill="white",color="black",binwidth=0.1) +
  facet_grid(Sample~.) +
  geom_vline(xintercept =log(1,2)) +
  geom_vline(xintercept = log(2,2)) +
  geom_vline(xintercept = log(3,2)) +
  theme_bw()
ggsave(paste0(path, "a4-count-hist.pdf"), width=5, height=9)
```
<br>    


#### Filter for strong peaks
```{r filter_peaks, eval=T}
library(dplyr)
# select peaks >= 3 rpm
pl3 <- lapply(pl, filter, rpm >=3)

# export each df as .csv file
lapply(names(pl3),
       function(x) write.csv(pl3[[x]], file=paste(paste0(path,x),"3rpm","csv",sep="."), row.names=F))

# convert each df to bed format
pl3bed <- lapply(pl3, function(x) {
  x <- x[,c("seqnames","start","end")]
})

# export each df as tdt bed file
lapply(names(pl3bed),
       function(x) write.table(pl3bed[x], file=paste(paste0(path,x),"3rpm","bed",sep="."), sep = "\t", 
                              quote=F, row.names=F,col.names=c("#chr","start","end")))
```
<br>    


#### Get genomic annotations of all peaks and strong peaks
+ are they biased for TSS?
+ HOMER annotatePeaks.pl
+ will write a script to loop over many bed files
+ for now ran each one-by-one
```{r annotate_pks, engine="bash"}
annotatePeaks.pl Spl_wVAT_Treg_rep1.3rpm.bed mm9 -annStats Spl_wVAT_Treg_rep1.3rpm.annStats.txt > Spl_wVAT_Treg_rep1.3rpm.annpks.txt

annotatePeaks.pl Spl_wVAT_Treg_rep1.bed mm9 -annStats Spl_wVAT_Treg_rep1.annStats.txt > Spl_wVAT_Treg_rep1.annpks.txt
```
<br> 


#### Plot distributions of genomic annotations before and after filtering
+ quick look in Excel
+ need to write code for ggplot
```{r annotate_dist, eval =T}
```
<br>    


#### Run pair-wise analyses for peaks with differential counts
+ DiffBind
```{r DiffBind}
library(DiffBind)

# create sample sheet
path <- "/Volumes/JD_WD2/JRD_CBDMLab/1_JRD_CBDM_PROJECTS/ProjectFolders/T-TregEpigenome/Data/1_Experimental/ATAC/Analyses/A4-vat-tconv-treg/"

Peaks <- list.files(paste0(path, "peaks/"))

# create columns for sample sheet
SampleID <- unlist(strsplit(Peaks, split = ".3rpm.bed"))
Tissue <- c(rep("Spleen",2),rep("VAT",4))   
Factor <- unlist(strsplit(SampleID, split = "_rep1"))
Factor <- unlist(strsplit(Factor, split = "_rep2"))
Condition <- "Treg"
#Treatment <- c("Batch1","Batch1","Batch2","Batch2","Batch1","Batch1")
Treatment <- c(rep("10wo",6))
Replicate <- c("1","2","1","2","1","2")
PeakCaller <- "homer"
PeakFormat <- "bed"
PeakPath <- list.files(paste0(path, "peaks/"), full.names = T)
bamPath <- list.files(paste0(path,"reads/"), full.names=T)

samples <- cbind(SampleID = SampleID, 
                 Tissue = Tissue, 
                 Factor = Factor, 
                 Condition = Condition,
                 Treatment = Treatment,
                 Replicate = Replicate,
                 bamReads = bamPath,
                 Peaks = PeakPath,
                 PeakCaller = PeakCaller,
                 PeakFormat = PeakFormat)
write.csv(samples, "samples.csv", row.names=F)
#Replicate <- sapply(SampleID, grep, "$")

##-----loading data-----------------------------------------------------------------------------------
# create DBA object
vat <- dba(sampleSheet = "samples.csv", bRemoveRandom = T, minOverlap = 2)

# make consensus peakset from replicates
vat <- dba.peakset(vat, consensus = -DBA_REPLICATE)

# make consensus peakset from samples and count reads
vat <- dba.count(vat, peaks = vat$masks$Consensus, score = DBA_SCORE_RPKM)

# save dba to file
dba.save(vat, file="vat", dir=path, pre="dba_", ext="RData", bMinimize=F)

# export consensus peakset with read counts
vat_con <- dba.peakset(vat, vat$masks$Consensus, bRetrieve = T, 
                       writeFile=paste0(path,"a4-consensus.rpkm.txt"), DataType=DBA_DATA_FRAME)
write.csv(vat_con, file=paste0(path, "a4-consensus.rpkm.csv"), row.names=F)

# save read count correlation heatmap
pdf(paste0(path,"a4.vat.reads.heat.pdf"), width=10, height=10, pagecentre = T)
par(oma = c(3,2,2,3))
dba.plotHeatmap(vat)
dev.off()

##-------------contrasts-----------------------------------------------------------------------------------
# establish contrast using factor (==Cell Type) metadata
contrast <- dba.contrast(vat, categories = DBA_FACTOR, minMembers=2)

# perform differential analysis
a1 <- dba.analyze(contrast) # [a]nalysis1

# return results of all contrasts
cr <- data.frame(dba.show(a1, bContrasts = T)) # [c]ontrast [r]esults
write.table(cr, file=paste0(path,"a4-contrasts.txt"), sep="\t", row.names=F)

# get peaks from specific contrasts
con4 <- dba.report(a1, contrast = 4, th=1, bUsePval = F, bNormalized =T, bCounts =T, initString = path, file="a4-con4.allpks")

# export plots from specific contrasts 
dba.plotMA(a1, contrast = 4, fold=0, bXY=T, th=0.01) # XY plot
dev.print(pdf, file=paste0(path,"a4-con4-xy.pdf")) 

dba.plotMA(a1, contrast = 4, fold=0, bXY=F, th=0.01) # MA plot
dev.print(pdf, file=paste0(path,"a4-con4-ma.pdf")) 

dba.plotHeatmap(a1, report=dba.report(a1, contrast=4, th = 0.01, bUsePval=F, fold=2), correlations =F,
                maxSites = 8000, bLog=T, score=DBA_SCORE_RPKM)
dev.print(pdf, file=paste0(path,"a4-con4-fdr01-fc2.pdf")) 

```
<br>    


#### Plot p-value vs fold-change
```{r plot_pval}
# read in datafile
path <- "/Volumes/JD_WD2/JRD_CBDMLab/1_JRD_CBDM_PROJECTS/ProjectFolders/T-TregEpigenome/Data/1_Experimental/ATAC/Analyses/A4-vat-tconv-treg/"
con4 <- read.csv(paste0(path,"_a4-con4.allpks.csv"), header = T)

# calculate number of peaks with FC >=2
Tconv.spec <- sum(con4$Fold >=2)
Treg.spec <- sum(con4$Fold <= -2)

# plot pval vs Fold-change
library(ggplot2)
g <- ggplot(data = con4, aes(x=Fold, y=log(p.value,10), color=FDR))
g + geom_point(size = 0.5) + 
  theme_bw() + 
  xlim(-10,10) + 
  xlab("Tconv / Treg") + 
  ylab("log10 P-value") + 
  annotate("text", x =7, y=-1, label=Tconv.spec, size =7, color = "red") +
  annotate("text", x=-7, y=-1, label=Treg.spec, size=7, color="red")
ggsave(filename=paste0(path,"a4-pval_vs_fold.pdf"), width=5,height=4)
```
<br>    

#### Export similar and differential peaks
```{r subset_pks}
Tconv.spec <- con4[con4$Fold >=2,]
Treg.spec <- con4[con4$Fold <= -2,]
Common <- con4[con4$Fold >-2 & con4$Fold <2,]

# add unique peak ids
n <- as.numeric(nrow(Tconv.spec))
prefix <- "Tconv_spec_pk_"
Tconv.spec$Name <- paste(prefix,seq(1:n), sep="")

n <- as.numeric(nrow(Treg.spec))
prefix <- "Treg_spec_pk_"
Treg.spec$Name <- paste(prefix,seq(1:n), sep="")

n <- as.numeric(nrow(Common))
prefix <- "Common_pk_"
Common$Name <- paste(prefix,seq(1:n), sep="")

# combine Treg and Tconv specific into all differential 
Alldiff <- rbind(Tconv.spec, Treg.spec)

# combine into list
dl <- list(Tconv_spec = Tconv.spec, Treg_spec = Treg.spec, Common = Common, All_diff = Alldiff) #[d]ifferential [l]ist

# convert each df to bed format
dl_bed <- lapply(dl, function(x) {
  x <- x[,c("Chr","Start","End","Name")]
})
 
# export bed files
lapply(names(dl_bed),
       function(x) write.table(dl_bed[x], file=paste(paste0(path,"a4-", x),"bed", sep="."), sep="\t",
                               quote=F, row.names=F, col.names=c("#chr","start","end","name")))
```
<br>    

#### Mapping peaks to genes with GREAT
+ Background = Whole genome background
+ Assembly = Mouse: NCBI build 37 (UCSC mm9, Jul 2007) 
+ Associate genomic regions = 
+ Basal+extension (constitutive 5.0 kb upstream and 1.0 kb downstream, up to 100.0 kb max extension). Curated regulatory domains are included.
+ overlap gene list for differential peaks with gene list for common peaks
+ did gene overlaps in R then used Excel vlookup to get peaks IDs for each gene
```{r great_genes}
# find overlaps between gene lists
adg <- read.table(file=paste0(path,"a4-Alldiff-great-geneID.txt")) # [a]ll [d]iff [g]enes
adg <- as.character(adg$V1)

acg <- read.table(file=paste0(path,"a4-Common-great-geneID.txt")) # [all] [c]ommon [genes]
acg <- as.character(acg$V1)

trsg <- read.table(file=paste0(path,"a4-Treg-spec-great-geneID.txt")) #[tr]eg-[s]pecific [g]enes
trsg <- as.character(trsg$V1)

tcsg <- read.table(file=paste0(path, "a4-Tconv-spec-great-geneID.txt"))
tcsg <- as.character(tcsg$V1)

overlap1 <- adg %in% acg # find differential that are also in common
overlap2 <- acg %in% adg # find common that are also in differential
overlap3 <- trsg %in% acg # find t-reg-specific that are also in common
overlap4 <- tcsg %in% acg # find t-conv-specific that are also in common

common_only <- acg[!overlap2] # check this again
diff_only <- adg[!overlap1]
comm_and_diff <- adg[overlap1]

trsg_with_common_pks <- trsg[overlap3] # [tr]eg[s]pecific[g]enes with common peaks
trsg_no_common_pks <- trsg[!overlap3] # [tr]eg[s]pecific[g]enes no common peaks

tcsg_with_common_pks <- tcsg[overlap4]
tcsg_no_common_pks <- tcsg[!overlap4]

write.table(trsg_with_common_pks, file=paste0(path, "a4-trsg_with_common_pks.txt"), sep = "\t",
            quote=F, row.names=F, col.names="GeneID")

write.table(trsg_no_common_pks, file=paste0(path, "a4-trsg_no_common_pks.txt"), sep = "\t",
            quote=F, row.names=F, col.names="GeneID")

# retrieve chromosomal coordinates for specific peaks that have common peak in same gene
peakID_trsg_with_common <- read.csv(file=paste0(path,"a4-trsg_with_common_pk.csv"), stringsAsFactors = F)
peakID_trsg_with_common <- peakID_trsg_with_common[-1]

non_zero <- lapply(peakID_trsg_with_common, function(x) x[x != "0"])
non_zero <- unlist(non_zero)
non_zero <- unique(non_zero)

overlap <- Treg.spec$Name %in% non_zero
  
Treg.spec.with.common <- Treg.spec[overlap,]

# retrieve chromosomal coordinates for specific peaks that do not have common peak in same gene
peakID_trsg_no_common <- read.csv(file=paste0(path,"a4-trsg_no_common_pk.csv"), stringsAsFactors = F)
peakID_trsg_no_common <- peakID_trsg_no_common[-1]

non_zero <- lapply(peakID_trsg_no_common, function(x) x[x != "0"])
non_zero <- unlist(non_zero)
non_zero <- unique(non_zero)

overlap <- Treg.spec$Name %in% non_zero

Treg.spec.no.common <- Treg.spec[overlap,]

# write both to bed files
# combine into list
pk_list <- list(Treg.spec.with.common = Treg.spec.with.common, Treg.spec.no.common = Treg.spec.no.common)
# convert each to bed
pk_list_bed <- lapply(pk_list, function(x) {
  x <- x[,c("Chr","Start","End","Name")]
})
# export bed files
lapply(names(pk_list_bed),
       function(x) write.table(pk_list_bed[x], file=paste(paste0(path,"a4-", x),"bed", sep="."), sep="\t",
                               quote=F, row.names=F, col.names=c("#chr","start","end","name")))
``` 
<br>



#### Motif-finding
+ use HOMER findMotifsGenome.pl
+ search all specific peaks, using whole genome Bg
+ repeat using splenic Bg "Spl_wVat_Treg_consensus_peaks.bed.txt"
+ search all common peaks same way
+ Params:
+ repeat-masked genome
+ size - given
`findMotifsGenome.pl a4-Treg_spec.bed mm9 a4-Treg-spec-motifsGiven/ -size given -mask -preparsedDir new/`
`findMotifsGenome.pl a4-Treg_spec.bed mm9 a4-Treg-spec-motifsGiven-splBg/ -size given -mask -bg Spl_wVat_Treg_consensus_peaks.bed.txt -preparsedDir new2/`





#### Clustering





