---
title: "A7-all-ttreg"
author: "JRD"
date: "March 7, 2016"
output: html_document
---

last modified: `r Sys.Date()`
<br>    

#### README
+ this document contains code and explanation for an analysis comparing Colon, Muscle, Splenic and VAT Treg
+ for details on how the input files used in this analysis were generated, see *ATAC_mapping.Rmd* and *ATAC_peaks.Rmd*
+ **difference between A5 and A7 is latter filters peaks by both min signal and replicate CV**
<br>    


#### R Markdown Specifications
Set global options
```{r setoptions, echo=F}
library(knitr)
opts_chunk$set(echo=F, eval=F, warning=F, message=F)
```
<br>    

Github
https://github.com/jdis24/ATAC.git
<br>    


#### JD's stat calculation functions
```{r stat_fxns}
# function to compute the number of peaks >=n rpm in each sample
# ARGS: df = dataframe of peak height rpms, peak_height_var = string of var name storing rpm values, cutoffs = numeric vector of rpm cutoffs 
# USAGE: number_peaks_above_nrpm(df, "peak_height_var", cutoffs)
number_peaks_above_nrpm <- function(df, rpm_var, cutoffs) { 
  # create indexing vector representing number of elements in cutoffs
  n <- length(cutoffs) 
  # create output vector whose length = length of cutoffs
  o <- numeric(n) 
  # create variable to store vector of extracted rpm values
  rpm <- df[[rpm_var]]
  # loop over each element of cutoffs and count number rows where peak height >= that element
  for(i in 1:n) { 
    o[i] <- as.numeric(sum(rpm >= cutoffs[[i]])) # calculate the total number of rows where rpm >= ith value in cutoffs  
  }
  # print output vector
  return(o) 
}


# define function to filter peaks and draw boxplot
# ARGS: df = df containing peak rpm values, colname == 'rpm'
#       n = rpm threshold
bp_nrpm <- function(df, n) {
  df %>%
    filter(rpm >= n) %>% # filter by rpm cutoff 
    ggplot(aes(x=Sample, y=log2_rpm, fill=Sample)) + # plot
    geom_boxplot(varwidth=T) +
    ylim(-2,10) +
    theme_bw() + 
    theme(axis.title.x = element_blank())
}


# calculate stats over all numeric columns of df 
row_stats <- function(df) {
  nums <- sapply(df, is.numeric) # extract numeric cols of df

  df_nums <- df[,nums] # subset df by numeric cols
  df_nums <- as.data.frame(apply(df_nums, 2, log,2)) # convert expression to log2

  stats <- transmute_(df_nums, 
                      Mean = ~rowMeans(df_nums),
                      SD = ~rowSds(as.matrix(df_nums)))
  CV <- stats$SD / stats$Mean

  df_other <- df[!nums]     
  cbind(df_other,stats,CV)
}

# calculate stats over any columns of df 
library(matrixStats)
row_stats2 <- function(df,v1,v2) {
  # v1 and v2 must put column numbe in quotes
  nums <- sapply(df, is.numeric) # extract numeric cols of df

  subset <- df[,v1:v2] # subset df by columns

  log2subset <- as.data.frame(apply(subset,2,log,2))

  means <- transmute_(subset,
                      mean = ~rowMeans(subset),
                      log2mean = ~log(rowMeans(subset),2),
                      log10mean = ~log(rowMeans(subset),10))
  meanlog2 <- transmute_(log2subset,
                         meanlog2 = ~rowMeans(log2subset))
  sd <- transmute_(subset,
                      SD = ~rowSds(as.matrix(subset)))
  sd2 <- transmute_(log2subset,
                      SD = ~rowSds(as.matrix(log2subset)))

  cv <- sd$SD / means$mean

  cv2 <- sd2$SD / meanlog2$mean

  #df_other <- df[!nums]
  cbind(PeakID = df$PeakID, subset,means,meanlog2,cv,cv2)
}
```
<br>  


#### JDs plotting functions
```{r plotting_fxns}
library(GGally)
jd_ggscatmat <- function(data, columns = 1:ncol(data), color = NULL, ...) {
  data.choose <- data[, columns]
  dn <- data.choose[sapply(data.choose, is.numeric)]
  if (ncol(dn) < 2) {
    stop("Not enough numeric variables to make a scatter plot matrix")
  }
  if (ncol(dn) == 0) 
    stop("All of your variables are factors. Need numeric variables to make scatterplot matrix.")
  a <- uppertriangle(data, columns = columns, color = color)
  if (is.null(color)) {
    plot <- jd_scatmat(data, columns = columns) + geom_text(data = a, 
                                                         aes_string(label = "r"), colour = "black")
  }
  else {
    plot <- jd_scatmat(data, columns = columns, color = color) + 
      geom_text(data = a, aes_string(label = "r", color = "colorcolumn")) + 
      labs(color = color)
  }
  factor <- data.choose[sapply(data.choose, is.factor)]
  if (ncol(factor) == 0) {
    return(plot)
  }
  else {
    warning("Factor variables are omitted in plot")
    return(plot)
  }
}

# draw scatterplot matrix
jd_scatmat <- function (data, columns = 1:ncol(data), color = NULL) 
{
  data.choose <- data[, columns]
  dn <- data.choose[sapply(data.choose, is.numeric)]
  if (ncol(dn) == 0) {
    stop("All of your variables are factors. Need numeric variables to make scatterplot matrix.")
  }
  else {
    ltdata.new <- lowertriangle(data, columns = columns, 
                                color = color)
    r <- ggplot(ltdata.new, mapping = aes_string(x = "xvalue", 
                                                 y = "yvalue")) + theme(axis.title.x = element_blank(), 
                                                                        axis.title.y = element_blank()) + facet_grid(ylab ~ 
                                                                                                                       xlab, scales = "free") + theme(aspect.ratio = 1)
    if (is.null(color)) {
      densities <- do.call("rbind", lapply(1:ncol(dn), 
                                           function(i) {
                                             data.frame(xlab = names(dn)[i], ylab = names(dn)[i], 
                                                        x = dn[, i])
                                           }))
      for (m in 1:ncol(dn)) {
        j <- subset(densities, xlab == names(dn)[m])
        r <- r + stat_density(aes(x = x, y = ..scaled.. * 
                                    diff(range(x)) + min(x)), data = j, position = "identity", 
                              geom = "line", color = "black")
      }
      r <- r + geom_point(na.rm = TRUE, size=0.5, color="dark green") + guides(fill=guide_legend(reverse=TRUE))
      return(r)
    }
    else {
      densities <- do.call("rbind", lapply(1:ncol(dn), 
                                           function(i) {
                                             data.frame(xlab = names(dn)[i], ylab = names(dn)[i], 
                                                        x = dn[, i], colorcolumn = data[, which(colnames(data) == 
                                                                                                  color)])
                                           }))
      for (m in 1:ncol(dn)) {
        j <- subset(densities, xlab == names(dn)[m])
        r <- r + stat_density(aes_string(x = "x", y = "..scaled.. * diff(range(x)) + min(x)", 
                                         colour = "colorcolumn"), data = j, position = "identity", 
                              geom = "line")
      }
      r <- r + geom_point(data = ltdata.new, aes_string(colour = "colorcolumn"), 
                          na.rm = TRUE, size=0.5, color="dark green") +
        guides(fill=guide_legend(reverse=TRUE))
      return(r)
    }
  }
}

```
<br>  


#### Pseudo
**call peaks and get read counts**
+ [these steps were done in "ATAC_peaks.Rmd"]
+ for each sample, call peaks in each replicate (homer)
+ then merge into single peakset (homer)
+ for each replicate, count reads in each peak, normalized to total library size (ChIPQC)
+ export .csv files of peaks and counts
+ [all subsequent steps done in this document]
**import peak files into peak list and export bed files**
+ import peaks and rpm-normalized counts from chipqc output
**generate table of peak numbers for range of rpm threshold values**
+ write function to compute number of peaks after applying a series of rpm cutoffs
+ boxplot each rpm cutoff
**plot distributions of read counts for each sample**
+ histograms
+ boxplots
+ before and after setting n rpm cutoff (start with >= 1rpm)
**filter for strong peaks**
+ plot distributions of read counts 
+ for different rpm values, examine peak shapes on browser from .bigwig files
+ filter for strong peaks (start with >=1 rpm) 
**TO DO: get genomic annotations of confident strong peaks** 
(compare to unfilterd to check degree of TSS enrichment)
**create merged peakset and get read counts for merged peaks**
+ use DiffBind to create merged peakset across all T-treg and to get read counts
+ export all peaks and their rpkm score in each sample 
**filter peaks by replicate CV**
+ import merged peakset with rpkm scores for each sample from DiffBind output
+ plot replicate rpkm values as scatter plots
+ plot mean vs cv for each set of replicates
+ use distributions to choose cv cutoff
+ filter on cv (start with cv <= 0.6)
+ plot replicate rpkm values post-filter
+ export all peaks and their stats for each sample
**check that filters worked on genome browser**
+ load .bigwig files for each sample with replicates
** quantile normalize read count scores**
+ start with rpm+cv-filtered peak set
+ plot read count distributions before qn
+ use bioconductor function to qn
+ plot read count distributions after qn
<br>  


#### Import peak files into peak list and export bed files
```{r peak_list, eval=T}
# create path to .csv files  
path <- "/Volumes/JD_WD2/JRD_CBDMLab/1_JRD_CBDM_PROJECTS/ProjectFolders/T-TregEpigenome/Data/1_Experimental/ATAC/Master_peaks/R/peaks_with_counts/working/"

# get filenames
filenames1 <- list.files(path = path, pattern = ".csv")
filenames <- paste(path, filenames1, sep="") # update filenames with path

# combine peak table .csv dfs into a list
colClass <- c(rep("factor",2), rep("numeric",3), "character", rep("numeric",4)) # set column classes
pl <- lapply(filenames, read.csv, colClasses = colClass) # [p]eak [l]ist

# name the list elements to match the files
names(pl) <- gsub("\\.csv", "", filenames1)

# convert each df to bed format
pl_bed <- lapply(pl, function(x) {
  x <- x[,c("seqnames","start","end")]
})

# export each df as tdt bed file
path <- "/Volumes/JD_WD2/JRD_CBDMLab/1_JRD_CBDM_PROJECTS/ProjectFolders/T-TregEpigenome/Data/1_Experimental/ATAC/Analyses/A7-all-ttreg/bed/"

lapply(names(pl_bed),
       function(x) write.table(pl_bed[x], file=paste(paste0(path,x),"bed",sep="."), sep = "\t", 
                              quote=F, row.names=F,col.names=c("#chr","start","end")))
```
<br>    


#### Generate table of peak numbers for range of rpm threshold values
```{r threshold_table, eval=T, results="asis"}
library(xtable)

# create cutoff vector of rpm values
cutoffs <- c(seq(0,5,by=0.5))

# apply cutoffs to list
peaks_pass_filter <- sapply(pl, number_peaks_above_nrpm, "rpm", cutoffs)

# transform list into dataframe 
ppf <- as.data.frame(peaks_pass_filter) #[p]eaks [p]ass [f]ilter
ppf <- cbind(ppf,"rpm_cutoff" = cutoffs) # add column for each cutoff

# write to file
path <- "/Volumes/JD_WD2/JRD_CBDMLab/1_JRD_CBDM_PROJECTS/ProjectFolders/T-TregEpigenome/Data/1_Experimental/ATAC/Analyses/A7-all-ttreg/plots/"
write.csv(ppf, file = paste0(path,"peak_totals_rpm_filtered.csv"), row.names=F)

# generate xtable
print(xtable(ppf), type="html", include.rownames=F)

# plot distributions of total peaks called at each cutoff
library(reshape2)
library(ggplot2)

ppfm <- melt(ppf, id.vars='rpm_cutoff',variable.name='no_peaks') # ppf[m]elted 
g <- ggplot(ppfm, aes(x=as.factor(rpm_cutoff),y=value))
g + geom_boxplot() +
  theme_bw()
ggsave(paste0(path, "a7-peakcut-boxplot.pdf"), width=7, height=5) # export pdf
```
<br>    


#### Plot distribution of read counts for each sample
```{r plots, eval=T}
##----- load libraries ----------------------------------------------------------------------------------------------
library(reshape2)
library(ggplot2)
library(dplyr)

##----- get peak list into shape for ggplot -------------------------------------------------------------------------

# set path for plots 
path <- "/Volumes/JD_WD2/JRD_CBDMLab/1_JRD_CBDM_PROJECTS/ProjectFolders/T-TregEpigenome/Data/1_Experimental/ATAC/Analyses/A7-all-ttreg/plots/"

# get peak list into shape for plotting
plm <- melt(pl, id.vars=names(pl[[1]])) # [p]eak [l]ist [m]elted
plm$L1 <- NULL # remove extraneous column

# calculcate log2 rpm values and append as new column
plm$log2_rpm <- log(plm$rpm, 2)

##----- find and filter out peaks where rpm =0 ---------------------------------------------------------------------

# show number of peaks where rpm =0 
rpm0pks <- plm$log2_rpm == "-Inf"
plm_rpm0pks <- plm[rpm0pks,]
nrow(plm_rpm0pks) # total pks rpm = 0

print("Peaks where rpm =0 ...") 
xtable(plm_rpm0pks)

# filter out peaks where rpm = 0
plmf <- plm[!rpm0pks,]  #plm[f]iltered

##----- draw boxplots before rpm filter -------------------------------------------------------------------------

# draw boxplots of all peaks in each sample
g <- ggplot(data = plmf, aes(x=Sample, y=log2_rpm, fill=Sample))
g + geom_boxplot(varwidth =T) +
  theme_bw() + 
  ylim(-2,10) +
  theme(axis.title.x = element_blank())
ggsave(paste0(path, "a7-boxplot.pdf"), width=7, height=5) # export pdf

# draw boxplot of peaks passing n rpm
# define function to filter peaks and draw plot
bp_nrpm <- function(df, n) {
  df %>%
    filter(rpm >= n) %>% # filter by rpm cutoff 
    ggplot(aes(x=Sample, y=log2_rpm, fill=Sample)) + # plot
    geom_boxplot(varwidth=T) +
    ylim(-2,10) +
    theme_bw() + 
    theme(axis.title.x = element_blank())
}

##----- draw boxplots post rpm filter -------------------------------------------------------------------------------

n <- 1 # set rpm cutoff
bp_nrpm(plmf, n) # draw plot
ggsave(paste0(path,"a7-boxplot-",n,"rpm-filtered.pdf"), width =7, height =5) # export pdf

# draw histograms of all peaks in each sample
g <- ggplot(data = plmf, aes(x=log2_rpm))
g + geom_histogram(fill="white",color="black",binwidth=0.1) +
  facet_grid(Sample~.) +
  geom_vline(xintercept =log(1,2)) +
  geom_vline(xintercept = log(2,2)) +
  geom_vline(xintercept = log(3,2)) +
  theme_bw()
ggsave(paste0(path, "a7-count-hist.pdf"), width=5, height=25)
```
<br>    


#### Filter for strong peaks
```{r filter_peaks, eval=T}
library(dplyr)
# select peaks >= nrpm
n <- 1
sp <- lapply(pl, filter, rpm >=n) #[s]trong[p]eaks

##----- export each filtered peakset as .csv file ------------------------------------------------------------------
path <- "/Volumes/JD_WD2/JRD_CBDMLab/1_JRD_CBDM_PROJECTS/ProjectFolders/T-TregEpigenome/Data/1_Experimental/ATAC/Analyses/A7-all-ttreg/1rpm-filtered/csv/"
suffix <- paste(n,'rpm','.csv',sep="")
lapply(names(sp),
       function(x) write.csv(sp[[x]], file=paste(path,x,'-',suffix,sep=""), row.names=F))


##----- export each filtered peakset as bed file ------------------------------------------------------------------
# convert each df to bed format
spbed <- lapply(sp, function(x) {
  x <- x[,c("seqnames","start","end")]
})

# export each df as tdt bed file
path <- "/Volumes/JD_WD2/JRD_CBDMLab/1_JRD_CBDM_PROJECTS/ProjectFolders/T-TregEpigenome/Data/1_Experimental/ATAC/Analyses/A7-all-ttreg/1rpm-filtered/bed/"

suffix <- paste(n, 'rpm','.bed',sep="")
lapply(names(spbed),
       function(x) write.table(spbed[x], file=paste(path,x,'-',suffix,sep=""), sep = "\t",
                               quote=F, row.names=F,col.names=c("#chr","start","end")))
```
<br>    


#### HOLD: Get genomic annotations of all peaks and strong peaks
+ are they biased for TSS?
+ HOMER annotatePeaks.pl
+ TODO
```{r annotate_pks, engine="bash"}
annotatePeaks.pl <filenmame> mm9 -annStats <filenmame>.annStats.txt > <filenmame>.annpks.txt

annotatePeaks.pl <filenmame>.bed mm9 -annStats <filenmame>.annStats.txt > <filenmame>.annpks.txt
```
<br>    


#### Create merged peak set and get normalized read counts 
+ DiffBind 
+ Used R 3.2.1 on Orchestra so didn't have to keep .bam locally
+ script called `a7.db.counts.R`
+ run with:
`module load stats/R/3.2.1`
`bsub -q priority -W 5:00 -R "rusage[mem=64000]" Rscript a7.db.counts.R`
```{r merge_and_count}
library(DiffBind)

##----- create sample sheet --------------------------------------------------------------------------------
path <- "/groups/cbdm-db/jrd26/ATAC_crossbatch/R/Analysis7/"

Peaks <- list.files(paste0(path, "bed/1rpm-filtered/"))

# create columns for sample sheet
SampleID <- unlist(strsplit(Peaks, split = "-1rpm.bed"))
Tissue <- c(rep('Colon',4),rep('Muscle',2),rep('Spleen',8),rep("VAT",2))
Factor <- unlist(strsplit(SampleID, split = "_rep1.2"))
Factor <- unlist(strsplit(Factor, split = "_rep1"))
Factor <- unlist(strsplit(Factor, split = "_rep2"))
Factor <- unlist(strsplit(Factor, split = "_rep3"))
Condition <- "Treg"
Treatment <- c('Batch1.2','Batch2','Batch1.2','Batch2',rep('Batch2',2),'Batch1','Batch2',
               'Batch1','Batch2',rep('Batch2',2),rep('Batch1',4))
Replicate <- rep(c('1','2'),8)
PeakCaller <- "homer"
PeakFormat <- "bed"
PeakPath <- list.files(paste0(path, "bed/1rpm-filtered/"), full.names = T)

path2 <- "/groups/cbdm-db/jrd26/ATAC_crossbatch/R/"
bamPath <- list.files(paste0(path2,"reads/"), full.names=T)

samples <- cbind(SampleID = SampleID, 
                 Tissue = Tissue, 
                 Factor = Factor, 
                 Condition = Condition,
                 Treatment = Treatment,
                 Replicate = Replicate,
                 bamReads = bamPath,
                 Peaks = PeakPath,
                 PeakCaller = PeakCaller,
                 PeakFormat = PeakFormat)
write.csv(samples, "samples.csv", row.names=F)


##-----loading data-----------------------------------------------------------------------------------
# create DBA object
ttreg <- dba(sampleSheet = "samples.csv", bRemoveRandom = T, minOverlap = 2)

# make consensus peakset from replicates
ttreg <- dba.peakset(ttreg, consensus = DBA_FACTOR)

# make consensus peakset from samples and count reads
ttreg <- dba.count(ttreg, peaks = ttreg$masks$Consensus, score = DBA_SCORE_RPKM)

# save dba to file
n <- 1  # save rpm filter value
dba.save(ttreg, file=paste0("a7-ttreg",n,'rpm-filtered'), dir=path, pre="dba_", ext="RData", bMinimize=F)

# export consensus peakset with read counts
# set filter
n <- 1
suffix <- paste('a7-ttreg-consensus-',n,'rpm-filtered-rpkm',sep="")
ttreg_con <- dba.peakset(ttreg, ttreg$masks$Consensus, bRetrieve = T, 
                         writeFile=paste0(path,suffix,'.txt'), DataType=DBA_DATA_FRAME)

write.csv(ttreg_con, file=paste0(path, suffix, '.csv'), row.names=F)

# save read count correlation heatmap
suffix <- paste('a7-ttreg-',n,'rpm-reads-heat',sep="")
pdf(paste0(path,suffix,'.pdf'), width=10, height=10, pagecentre = T)
par(oma = c(3,2,2,3))
dba.plotHeatmap(ttreg)
dev.off()

# save read count PCA
suffix <- paste('a7-ttreg-',n,'rpm-pca',sep="")
pdf(paste0(path,suffix,'.pdf'), width=10, height=10, pagecentre = T)
par(oma = c(3,2,2,3))
dba.plotPCA(ttreg, attributes=DBA_TISSUE, label=DBA_FACTOR, score=DBA_SCORE_RPKM)
dev.off()

```
<br>    


#### Filter peaks by replicate CV
+ import results of DiffBind counts
+ **give each peak a unique ID**
+ make separate df for each sample with replicate rpkm, mean, log2mean, cv , cvlog2 (cv2)
+ combine dfs into list
+ melt list 
+ ggplot mean vs cv2 for each sample
+ look at different cv cutoffs on browser
+ create function to filter by cv
+ apply function over sample list
+ recombine into single df for each peak, using peakID 
```{r cv_filter}

##--- import peak file with normalized counts -------------------------------------------------------------
path <- "/Volumes/JD_WD2/JRD_CBDMLab/1_JRD_CBDM_PROJECTS/ProjectFolders/T-TregEpigenome/Data/1_Experimental/ATAC/Analyses/A7-all-ttreg/1rpm-filtered/csv/"
ttreg <- read.csv(file = paste(path,'a7-ttreg-consensus-1rpm-filtered-rpkm.csv',sep=""), header =T)
summary(ttreg)

##----plot read count distributions after merge -------------------------------------------------------------
# set plots path
path <- "/Volumes/JD_WD2/JRD_CBDMLab/1_JRD_CBDM_PROJECTS/ProjectFolders/T-TregEpigenome/Data/1_Experimental/ATAC/Analyses/A7-all-ttreg/1rpm-filtered/plots/"

# set file suffix
n <-1
suffix <- paste0('a7-ttreg-', n,'rpm-filtered-merged-counts-rpkm')

# plot and save
pdf(paste0(path,suffix,'.pdf'), width=8, height=6)
par(oma = c(4,1,0.5,1))
h <- 2 # set height of horizontal line
boxplot(log(ttreg[4:19],2), las = 2, cex.axis = 0.7, 
        ylab='log2 rpkm', 
        main=paste('merged atac peaks ','n=',nrow(ttreg),'; abline rpm=',h,sep=""))
abline(h=h, col = 'red')
dev.off()


##---- give each peak unique ID --------------------------------------------------------------------------------
peakName <- paste('peak_', 1:nrow(ttreg),sep="")
ttreg <- cbind(PeakID = peakName, ttreg)
# export 
path <- "/Volumes/JD_WD2/JRD_CBDMLab/1_JRD_CBDM_PROJECTS/ProjectFolders/T-TregEpigenome/Data/1_Experimental/ATAC/Analyses/A7-all-ttreg/1rpm-filtered/csv/"
write.csv(ttreg, file = paste(path,'a7-ttreg-consensus-',n,'rpm-filtered-rpkm-named.csv',sep=""), row.names=F)

## start here if experimenting with different CV
# load
path <- "/Volumes/JD_WD2/JRD_CBDMLab/1_JRD_CBDM_PROJECTS/ProjectFolders/T-TregEpigenome/Data/1_Experimental/ATAC/Analyses/A7-all-ttreg/1rpm-filtered/csv/"
ttreg <- read.csv(paste0(path,'a7-ttreg-consensus-1rpm-filtered-rpkm-named.csv'))
ttreg <- ttreg[-2] # remove extraneous column

##---- apply row stats functions to each sample ----------------------------------------------------------------
names(ttreg) # view column names

# create new df for each sample's stats
library(dplyr)
Cln_Nrp1n <- row_stats2(ttreg, "5","6")
Cln_Nrp1p <- row_stats2(ttreg, "7","8")
Mus <- row_stats2(ttreg, "9","10")
Spl_wCln_Nrp1n <- row_stats2(ttreg, "11","12")
Spl_wCln_Nrp1p <- row_stats2(ttreg, "13","14")
Spl_wMus <- row_stats2(ttreg,"15","16")
Spl_wVat <- row_stats2(ttreg,"17","18")
Vat <- row_stats2(ttreg, "19","20")

# combine the dfs into a list
psl <- list(Cln_Nrp1n, Cln_Nrp1p, Mus, Vat, Spl_wCln_Nrp1n, Spl_wCln_Nrp1p, Spl_wMus, Spl_wVat) #[p]eak[s]tat[l]ist
# name each element of the list
psl.names <- c('Cln_Nrp1n', 'Cln_Nrp1p', 'Mus', 'Vat', 'Spl_wCln_Nrp1n', 'Spl_wCln_Nrp1p', 
                'Spl_wMus', 'Spl_wVat') 
names(psl) <- psl.names
names(psl) # check

# rename the columns of each df
stat_names <- c('PeakID','Rep1_rpkm', 'Rep2_rpkm','Mean','log2_mean', 'log10_mean', 'Mean_of_log2_rpkm',
                'CV','CV_of_log2_rpkm')

for (i in seq_along(psl)) {
  colnames(psl[[i]]) <- stat_names
}
# check 
head(psl[[1]])

# create sample IDs
samples <- as.list(psl.names)
sampleID <- lapply(samples, function(x) {rep(x,nrow(ttreg))})
# append sampleID to each df
ttreg.stats <- list()
for (i in 1:length(psl)) {
  ttreg.stats[[i]] <- cbind(SampleID = sampleID[[i]], psl[[i]])
}
# name list elements
names(ttreg.stats) <- psl.names

# check
lapply(ttreg.stats, head)


##----plot replicate mean and cvs --------------------------------------------------------------------------------
library(reshape2)
library(ggplot2)

# set plots path
path <- "/Volumes/JD_WD2/JRD_CBDMLab/1_JRD_CBDM_PROJECTS/ProjectFolders/T-TregEpigenome/Data/1_Experimental/ATAC/Analyses/A7-all-ttreg/1rpm-filtered/plots/"
# set rpm filter for file suffix
n <- 1

# get into shape for ggplot
slm <- melt(ttreg.stats, id.vars = colnames(ttreg.stats[[1]])) #[s]tat[l]ist[m]elted
slm$L1 <- NULL

# plot replicates against each other
p <- ggplot(slm, aes(x=log(Rep1_rpkm,2),y=log(Rep2_rpkm,2)))
p + geom_point(size= 0.7, shape = 1, alpha=0.6) +
  facet_wrap(~SampleID) + 
  theme_bw()
ggsave(paste(path,'a7-',n,'rpm-rep-scatter-prefilter.png',sep=""), width=8, height=8)

# plot mean vs cv
p <- ggplot(slm, aes(x=log2_mean, y=CV))
p + geom_point(size= 0.7, shape = 1, alpha=0.6) +
  facet_wrap(~SampleID) + 
  theme_bw()
ggsave(paste(path,'a7-',n,'rpm-rep-cv-prefilter.png',sep=""), width=8, height=8)


##---- filter by max cv  --------------------------------------------------------------------------------------
# filter by cv <= 0.8
library(dplyr)
n <- 0.8 # set cv filter value
fbcv.list <- lapply(ttreg.stats, filter, CV <= n) # [f]iltered[b]y[cv]

names(fbcv.list) <- psl.names  # name list elements (recycle previous names)
lapply(fbcv.list, head) # check
  
# take a look at results
prefilter <- unlist(lapply(ttreg.stats, nrow))
postfilter <- unlist(lapply(fbcv.list, nrow))
cvfilteredtab <- rbind(prefilter, postfilter)
cvfilteredtab

# get list into shape for plotting
fbcvm <- melt(fbcv.list, id.vars = colnames(fbcv.list[[1]])) #[fbcvm]elted
fbcvm$L1 <- NULL


## --- plot after cv filter ---------------------------------------------------------------------------------------
path <- "/Volumes/JD_WD2/JRD_CBDMLab/1_JRD_CBDM_PROJECTS/ProjectFolders/T-TregEpigenome/Data/1_Experimental/ATAC/Analyses/A7-all-ttreg/1rpm-cv08/plots/"
n <- 1 # set rpm filter for filename suffix
c <- 0.8 # set cv filter for filename suffix

# distribution of read counts 
p <- ggplot(fbcvm, aes(x=log2_mean))
p + geom_histogram(fill="white",color="black",binwidth=0.1) +
  facet_grid(SampleID~.) +
  geom_vline(xintercept =log(1,2)) +
  geom_vline(xintercept = log(2,2)) +
  geom_vline(xintercept = log(3,2)) +
  theme_bw()
ggsave(paste(path,'a7-',n,'rpm-',c,'cv','-rep-rpkm-hist.pdf',sep=""), width=8, height=10)
 
# plot replicates against each other after cv filter
p <- ggplot(fbcvm, aes(x=log(Rep1_rpkm,2),y=log(Rep2_rpkm,2)))
p + geom_point(size= 0.7, shape = 1, alpha=0.6) +
  facet_wrap(~SampleID) + 
  theme_bw()
ggsave(paste(path,'a7-',n,'rpm-',c,'cv','-rep-scatter-postfilter.png',sep=""), width=8, height=8)


## --- lookup and paste coordinates for each peak -----------------------------------------------------------------
library(dplyr)

# join filtered peak files to original ttreg df 
fbcv.list.bed <- lapply(fbcv.list, function(x) { left_join(x,ttreg[1:4],by="PeakID") } )
# check for NAs
sapply(fbcv.list.bed, function(x) all(is.na(x)))
# check
lapply(fbcv.list.bed, head)
# re-arrange columns of each df
fbcv.list.bed <- lapply(fbcv.list.bed, function(x) { cbind(x[1:2],x[11:13],x[3:10])})
names(fbcv.list.bed) <- psl.names  # name list elements (recycle previous names)
lapply(fbcv.list.bed, head)

## --- export cv-filtered list and tables for each sample ----------------------------------------------------------

# save list of cv-filtered dfs
path <- "/Volumes/JD_WD2/JRD_CBDMLab/1_JRD_CBDM_PROJECTS/ProjectFolders/T-TregEpigenome/Data/1_Experimental/ATAC/Analyses/A7-all-ttreg/1rpm-cv08/RDa/"
n <- 1 # set rpm filter for filename suffix
c <- 0.8 # set cv filter for filename suffix
save(fbcv.list.bed, file = paste(path,'a7-ttreg-',n,'rpm-',c,'cv','-list.RDa',sep=""))
# load
load(paste0(path,'a7-ttreg-1rpm-0.8cv-list.RDa'))

# export each df as .csv file
path <- "/Volumes/JD_WD2/JRD_CBDMLab/1_JRD_CBDM_PROJECTS/ProjectFolders/T-TregEpigenome/Data/1_Experimental/ATAC/Analyses/A7-all-ttreg/1rpm-cv08/csv/"
suffix <- paste('-',n,'rpm-',c,'cv','-filtered','.csv',sep="")
lapply(names(fbcv.list.bed), 
       function(x) write.csv(fbcv.list.bed[[x]], file=paste(path,x,suffix,sep=""), row.names=F))


##---- merge all the filtered peaks ------------------------------------------------------------------------------
# set path
path <- "/Volumes/JD_WD2/JRD_CBDMLab/1_JRD_CBDM_PROJECTS/ProjectFolders/T-TregEpigenome/Data/1_Experimental/ATAC/Analyses/A7-all-ttreg/1rpm-cv08/csv/"

# get filenames
filenames <- list.files(path = path, pattern = ".csv", full.names=T)

# combine into a list
colClass <- c('factor','character','factor',rep('integer',2),rep('numeric',8))
pl <- lapply(filenames, read.csv, colClasses = colClass) # [p]eak[l]ist

# name list elements to match filenames 
filenames1 <- list.files(path=path, pattern=".csv")
sample.names <- strsplit(filenames1, split='-1rpm-')
sample.names <- unlist(lapply(sample.names, function(x) x[[1]]))
names(pl) <- sample.names
names(pl) # check
lapply(pl, head)

# remove extraneous columns
pl <- lapply(pl, function(x) x[1:7])
str(pl[[1]])

# create vector of new column names for rpkm columns
r1 <- c('_rpkm_r1')
r2 <- c('_rpkm_r2')
rpkm1 <- c(paste(sample.names,r1,sep=""),paste(sample.names,r2,sep=""))
rpkm1 <- rpkm1[order(rpkm1)]

# rename rpkm dfs 
namelist <- list()
for (i in 1:length(names(pl))) {
  ext <- c('_rpkm_r1','_rpkm_r2')
  namelist[[i]] <- paste(names(pl)[[i]],ext,sep="")
}

namelist2 <- list()
for (i in 1:length(names(pl))) {
  names1 <- colnames(pl[[1]][1:5])
  namelist2[[i]] <- c(names1, namelist[[i]])
}


for (i in seq_along(pl)) {
  colnames(pl[[i]]) <- namelist2[[i]]
}

lapply(pl, head) # check 
pl <- lapply(pl, function(x) {x[-1]}) # remove SampleID
lapply(pl, head) # check 

# extract individual df
list2env(pl, .GlobalEnv)

# merge
hcp <- inner_join(Cln_Nrp1n,Cln_Nrp1p, by='PeakID') #[h]igh[c]onfidence[p]eaks
hcp <- inner_join(hcp, Mus, by='PeakID')
hcp <- inner_join(hcp, Spl_wCln_Nrp1n, by='PeakID')
hcp <- inner_join(hcp, Spl_wCln_Nrp1p, by = 'PeakID')
hcp <- inner_join(hcp, Spl_wMus, by= 'PeakID')
hcp <- inner_join(hcp, Spl_wVat, by='PeakID')
hcp <- inner_join(hcp, Vat, by = 'PeakID')

# remove extraneous columns
chr <- grepl(pattern = c("CHR*") ,colnames(hcp))
start <- grepl(pattern = c("START*"), colnames(hcp))
end <- grepl(pattern = c("END*"), colnames(hcp))
hcp <- hcp[, !(chr | start | end)]


##--- add peak coordinaates ----------------------------------------------------------------------------------------
# import peak file with normalized counts -
path <- "/Volumes/JD_WD2/JRD_CBDMLab/1_JRD_CBDM_PROJECTS/ProjectFolders/T-TregEpigenome/Data/1_Experimental/ATAC/Analyses/A7-all-ttreg/1rpm-filtered/csv/"
ttreg <- read.csv(file = paste(path,'a7-ttreg-consensus-1rpm-filtered-rpkm-named.csv',sep=""), header =T)
names(ttreg)
# remove extraneous column
ttreg <- ttreg[-2]
# join to add peak coords
hcp.df <- left_join(hcp, ttreg[1:4], by = "PeakID") # hcp.[d]ata[f]rame
# rearrange columns
hcp.df <- cbind(hcp.df[1],hcp.df[18:20],hcp.df[2:17])
names(hcp.df)


## --- save -------------------------------------------------------------------------------------------------------
n <- 1 # set rpm filter for filename suffix
c <- 0.8 # set cv filter for filename suffix

# RDa
path <- "/Volumes/JD_WD2/JRD_CBDMLab/1_JRD_CBDM_PROJECTS/ProjectFolders/T-TregEpigenome/Data/1_Experimental/ATAC/Analyses/A7-all-ttreg/1rpm-cv08/RDa/"
save(hcp.df, file = paste(path,'a7-ttreg-',n,'rpm-',c,'cv-filtered-merged','.RDa',sep=""))

# csv
path <- "/Volumes/JD_WD2/JRD_CBDMLab/1_JRD_CBDM_PROJECTS/ProjectFolders/T-TregEpigenome/Data/1_Experimental/ATAC/Analyses/A7-all-ttreg/1rpm-cv08/csv/"
write.csv(hcp.df, file=paste(path,'a7-ttreg-',n,'rpm-',c,'cv-filtered-merged','.csv',sep="" ), row.names=F)


## --- plot -----------------------------------------------------------------------------------------------------
# set path
path <- "/Volumes/JD_WD2/JRD_CBDMLab/1_JRD_CBDM_PROJECTS/ProjectFolders/T-TregEpigenome/Data/1_Experimental/ATAC/Analyses/A7-all-ttreg/1rpm-cv08/plots/"
n <- 1 # set rpm filter for filename suffix
c <- 0.8 # set cv filter for filename suffix

# scatter matrices to get idea of pair-wise FC
library(GGally)
# make new df with log2 values
hcp.log2 <- data.frame(lapply(hcp.df[5:20], log,2))

# plot all ttreg
jd_ggscatmat(hcp.log2[c(1:6,15:16)])
# save
ggsave(paste(path,'a7-',n,'rpm-',c,'cv-filtered-ttreg-rpkm.png',sep=""), width=11, height=11)

# colon vs matching spleen
jd_ggscatmat(hcp.log2[c(1:4,7:10)])
# save
ggsave(paste(path,'a7-',n,'rpm-',c,'cv-filtered-cln-spln-rpkm.png',sep=""), width=14, height=14)

# mus vs matching spleen
jd_ggscatmat(hcp.log2[c(5:6,11:12)])
# save
ggsave(paste(path,'a7-',n,'rpm-',c,'cv-filtered-mus-spln-rpkm.png',sep=""), width=8, height=8)

# vat vs matching spleen
jd_ggscatmat(hcp.log2[c(13:16)])
# save 
ggsave(paste(path,'a7-',n,'rpm-',c,'cv-filtered-vat-spln-rpkm.png',sep=""), width=8, height=8)
```
<br>  


#### Quantile normalize read count values
**Pseudo:**  
+ Plot distributions of read counts in the merged peak set for all samples 
+ Quantile normalize using normalize.quantiles() (preprocessCore)
+ plot distributions of read counts in the merged peak set post QN
```{r qn}
library(reshape2)
library(ggplot2)

## --- import file --------------------------------------------------------------------------------------------------

path <- "/Volumes/JD_WD2/JRD_CBDMLab/1_JRD_CBDM_PROJECTS/ProjectFolders/T-TregEpigenome/Data/1_Experimental/ATAC/Analyses/A7-all-ttreg/1rpm-cv08/csv/"
hcp <- read.csv(file = paste(path,'a7-ttreg-1rpm-0.8cv-filtered-merged.csv',sep=""), header =T)


## --- plot read count distributions after merge --------------------------------------------------------------------
# set plots path
path <- "/Volumes/JD_WD2/JRD_CBDMLab/1_JRD_CBDM_PROJECTS/ProjectFolders/T-TregEpigenome/Data/1_Experimental/ATAC/Analyses/A7-all-ttreg/1rpm-cv08/plots/"

# set file suffix
n <- 1 # set rpm filter for filename suffix
c <- 0.8 # set cv filter for filename suffix
suffix <- paste0('a7-ttreg-', n,'rpm-',c,'cv-filtered-merged-counts-rpkm')

# plot and save
pdf(paste0(path,suffix,'.pdf'), width=8, height=6)
par(oma = c(4,1,0.5,1))
h <- 2 # set height of horizontal line
boxplot(log(hcp[5:20],2), las = 2, cex.axis = 0.7, 
        ylab='log2 rpkm', 
        main=paste('merged, rpm+cv-filtered atac peaks ','n=',nrow(hcp),'; abline rpm=',h,sep=""))
abline(h=h, col = 'red')
dev.off()


## --- quantile normalization ---------------------------------------------------------------------------------------
library(preprocessCore)

# create matrix of rpm values
hcpm <- as.matrix(hcp[,5:20]) # [h]igh[c]onfidence[p]eak[m]atrix

# qn using preprocess core
hcpqn.mat <- normalize.quantiles(hcpm) #hcp[q]uantile[n]ormalized.[mat]rix

# set file suffix
n <- 1 # set rpm filter for filename suffix
c <- 0.8 # set cv filter for filename suffix
suffix <- paste0('a7-ttreg-', n,'rpm-',c,'cv-filtered-merged-counts-rpkm-qn')

# set col names 
colnames(hcpqn.mat) <- names(hcp[5:20])

## --- plot read count distributions after merge then qn ---------------------------------------------------
path <- "/Volumes/JD_WD2/JRD_CBDMLab/1_JRD_CBDM_PROJECTS/ProjectFolders/T-TregEpigenome/Data/1_Experimental/ATAC/Analyses/A7-all-ttreg/1rpm-cv08/plots/"

# plot and save
pdf(paste0(path,suffix,'.pdf'), width=8, height=6)
par(oma = c(4,1,0.5,1))
h <- 2 # set height of horizontal line
boxplot(log(hcpqn.mat,2), las = 2, cex.axis = 0.7, 
        ylab='log2 rpkm', 
        main=paste('merged, rpm+cv-filtered atac peaks ','n=',nrow(hcp),'; abline rpm=',h,sep=""))
abline(h=h, col = 'red')
dev.off()


## --- export qn rpkm values -------------------------------------------------------------------------------
path <- "/Volumes/JD_WD2/JRD_CBDMLab/1_JRD_CBDM_PROJECTS/ProjectFolders/T-TregEpigenome/Data/1_Experimental/ATAC/Analyses/A7-all-ttreg/1rpm-cv08/RDa/"
# set file suffix
n <- 1 # set rpm filter for filename suffix
c <- 0.8 # set cv filter for filename suffix
suffix <- paste0('a7-ttreg-', n,'rpm-',c,'cv-filtered-merged-counts-rpkm-qn-mat')

save(hcpqn.mat, file=paste0(path,suffix,'.RDa'))

# as df with chrom info pasted on
# add chrom coordinates info
hcpqn.df <- cbind(hcp[1:4],hcpqn.mat)

suffix <- paste0('a7-ttreg-', n,'rpm-',c,'cv-filtered-merged-counts-rpkm-qn-df')
save(hcpqn.df, file=paste0(path,suffix,'.RDa'))
```
<br>  


#### Average repliate values and generate read count matrix to use for clustering
**Pseudo:**  
+ Take average of replicate values for each sample
+ Plot distributions of read counts after averaging reps
+ **Plot PCA after averaging reps** (TO DO)
+ Convert replicate averages to log2
+ **Give each peak a unique ID**
+ Export avg rpkm, log2rpkm of avgs 
```{r count_mat}
##---- # load qn read count matrix and df ------------------------------------------------------------------------
n <- 1
c <- 0.8
path <- "/Volumes/JD_WD2/JRD_CBDMLab/1_JRD_CBDM_PROJECTS/ProjectFolders/T-TregEpigenome/Data/1_Experimental/ATAC/Analyses/A7-all-ttreg/1rpm-cv08/RDa/"

# read count matrix
#suffix <- paste0('a7-ttreg-', n,'rpm-',c,'cv-filtered-merged-counts-rpkm-qn-mat')
#load(paste0(path,suffix,'.RDa'))

# read count df
suffix <- paste0('a7-ttreg-', n,'rpm-',c,'cv-filtered-merged-counts-rpkm-qn-df')
load(paste0(path,suffix,'.RDa'))


##----avg replicates  -------------------------------------------------------------------------------------
library(dplyr)

# compute replicate averages 
ttr.avg <- hcpqn.df %>% # ttreg averages
  transmute(mean1 = rowMeans(.[5:6]),
            mean2 = rowMeans(.[7:8]),
            mean3 = rowMeans(.[9:10]),
            mean4 = rowMeans(.[11:12]),
            mean5 = rowMeans(.[13:14]),
            mean6 = rowMeans(.[15:16]),
            mean7 = rowMeans(.[17:18]),
            mean8 = rowMeans(.[19:20]))

# generate new column names
samples <- names(hcpqn.df[-c(1:4)])
samples <- as.character(strsplit(samples, split = '_rep1.2'))
samples <- as.character(strsplit(samples, split='_rep[1,2,3]'))
samples <- as.character(strsplit(samples, split='_rpkm_r[1,2]'))
samples <- unique(samples)

names(ttr.avg) <- samples # rename columns
strand <- '+' # add strand

ttr.avg <- cbind(hcpqn.df[1:4],Strand = strand, ttr.avg) # bind together

# re-order to put all t-treg near each other, then all splenic
ttr.avg <- ttr.avg[c(1:8,13,9:12)]

##---- export -------------------------------------------------------------------------------------------------
n <- 1
c <- 0.8
path <- "/Volumes/JD_WD2/JRD_CBDMLab/1_JRD_CBDM_PROJECTS/ProjectFolders/T-TregEpigenome/Data/1_Experimental/ATAC/Analyses/A7-all-ttreg/1rpm-cv08/RDa/"

# raw values as RDa
suffix <- paste0('a7-ttreg-',n,'rpm-',c,'cv-filtered-rep-avg-counts-rpkm-qn')
save(ttr.avg, file=paste0(path,suffix,'.RDa'))

# load 
load(paste0(path, 'a7-ttreg-1rpm-0.6cv-filtered-rep-avg-counts-rpkm-qn.RDa'))

# raw values as .csv
path <- "/Volumes/JD_WD2/JRD_CBDMLab/1_JRD_CBDM_PROJECTS/ProjectFolders/T-TregEpigenome/Data/1_Experimental/ATAC/Analyses/A7-all-ttreg/1rpm-cv08/csv/"
write.csv(ttr.avg, file=paste0(path,suffix,'.csv'), row.names=F) # as csv

# as .bed
path <- "/Volumes/JD_WD2/JRD_CBDMLab/1_JRD_CBDM_PROJECTS/ProjectFolders/T-TregEpigenome/Data/1_Experimental/ATAC/Analyses/A7-all-ttreg/1rpm-cv08/bed/"
write.table(ttr.avg[1:5],file=paste0(path,suffix,'.bed'), row.names=F, sep="\t",
            col.names=c('#PeakID','Chr','Start','End','Strand'))


##---- plot -------------------------------------------------------------------------------------------------
# set filenames
path <- "/Volumes/JD_WD2/JRD_CBDMLab/1_JRD_CBDM_PROJECTS/ProjectFolders/T-TregEpigenome/Data/1_Experimental/ATAC/Analyses/A7-all-ttreg/1rpm-cv08/plots/" # set plots path
n <- 1
c <- 0.8
suffix <- paste0('a7-ttreg-', n,'rpm-',c,'cv-filtered-rep-avg-counts-rpkm-qn') # set file suffix

# boxplot average read count distributions
pdf(paste0(path,suffix,'.pdf'), width=8,height=6, pagecentre = T)
par(oma = c(4,1,0.5,1))
boxplot(log(ttr.avg[6:13],2), las=2, cex.axis=0.7,
        ylab='replicate avg log2rpkm',main=paste('merged rpm and cv-filtered atac peaks','n=',nrow(ttr.avg),sep=""))
dev.off()


##---- log2 convert and export --------------------------------------------------------------------------------
ttr.avg.l2 <- ttr.avg #[t][tr]eg[a]verage[l]og[2]
ttr.avg.l2[6:13] <- as.data.frame(lapply(ttr.avg.l2[6:13], log,2))


# export
path <- "/Volumes/JD_WD2/JRD_CBDMLab/1_JRD_CBDM_PROJECTS/ProjectFolders/T-TregEpigenome/Data/1_Experimental/ATAC/Analyses/A7-all-ttreg/1rpm-cv08/csv/" # set data path
n <- 1
c <- 0.8
suffix <- paste0('a7-ttreg-', n,'rpm-',c,'cv-filtered-rep-avg-counts-log2-rpkm-qn')
# csv
write.csv(ttr.avg.l2, file=paste0(path,suffix,'.csv'), row.names=F)

# RDa
path <- "/Volumes/JD_WD2/JRD_CBDMLab/1_JRD_CBDM_PROJECTS/ProjectFolders/T-TregEpigenome/Data/1_Experimental/ATAC/Analyses/A7-all-ttreg/1rpm-cv08/RDa/"  # set data path
save(ttr.avg.l2, file=paste0(path,suffix,'.RDa'))
```
<br>


#### For qn-rpm matrix of peak scores, convert peakIDs to row names
+ (would like to use GeneName or Functional Element but rownames have to be unique) 
```{r peakID_rownames}
# set path and filename
#path <- "/Volumes/JD_WD2/JRD_CBDMLab/1_JRD_CBDM_PROJECTS/ProjectFolders/T-TregEpigenome/Data/1_Experimental/ATAC/Analyses/A7-all-ttreg/1rpm-cv06/bed/"
#filename <- 'a7-ttreg-1rpm-0.6cv-filtered-rep-avg-counts-rpkm-qn.bed'
# read in
#ttr.annot <- read.delim(paste0(path,filename), header=T, check.names=T, sep='\t')
# change col names
#names(ttr.annot) <- c('PeakID', 'Chr','Start','End','Strand')

# load counts 
path <- "/Volumes/JD_WD2/JRD_CBDMLab/1_JRD_CBDM_PROJECTS/ProjectFolders/T-TregEpigenome/Data/1_Experimental/ATAC/Analyses/A7-all-ttreg/1rpm-cv08/RDa/" 
filename <- 'a7-ttreg-1rpm-0.8cv-filtered-rep-avg-counts-log2-rpkm-qn.RDa'
load(paste0(path,filename))


## --- convert log2rpm df to matrix with rownames == PeakID -----------------------------------------------------
library(dplyr)
# convert peakID to rownames
rownames(ttr.avg.l2) <- ttr.avg.l2$PeakID
# remove extraneous columns
ttr.avg.l2 <- ttr.avg.l2[-c(1:5)]
# convert to matrix
ttrm <- as.matrix(ttr.avg.l2)
# check
head(attributes(ttrm)[[1]])
head(dimnames(ttrm)[[1]],5)
dimnames(ttrm)[[2]]


##---- export -------------------------------------------------------------------------------------------------
path <- "/Volumes/JD_WD2/JRD_CBDMLab/1_JRD_CBDM_PROJECTS/ProjectFolders/T-TregEpigenome/Data/1_Experimental/ATAC/Analyses/A7-all-ttreg/1rpm-cv08/RDa/" # set data path
n <- 1 # set rpm cutoff for filenames
c <- 0.8 # set cv cutoff for filenames

# as RDa
suffix <- paste0('a7-ttreg-', n, 'rpm-',c,'cv-filtered-rep-avg-counts-log2-rpkm-qn-peakID-mat')
save(ttrm, file=paste0(path,suffix,'.RDa'))

# as csv
path <- "/Volumes/JD_WD2/JRD_CBDMLab/1_JRD_CBDM_PROJECTS/ProjectFolders/T-TregEpigenome/Data/1_Experimental/ATAC/Analyses/A7-all-ttreg/1rpm-cv08/csv/" # set data path
write.csv(ttrm, file = paste0(path,suffix,'.csv'))
```
<br>  


#### Finding best heatmap colors for clusters
```{r colors}
eqSpect <- colorRampPalette(
               c("#f2003c", "#F66900", "#F19100", "#F1B100",
                 "#EFD300", "#f0ea00", "#CBDB00", "#9DD501",
                 "#5ED108", "#00AF63", "#00A78E", "#00a3ac",
                 "#0093af", "#0082b2", "#006ebf", "#4F37C2",
                 "#8000D3", "#A001BF", "#BA00A4", "#D0007F"),
                 space="rgb",
                 interpolate="spline")
n <- 16

niceCols <- eqSpect(n)

# or 
niceCols <- colorRampPalette(rev(brewer.pal(10, 'Spectral')))(50)
niceCols <- rev(brewer.pal(11, 'Spectral'))
```
<br>  


#### Divide peaks into similar and variable subsets
```{r find_var_peaks}
# --- read in df of peak scores ---------------------------------------------------------------------------------
path <- "/Volumes/JD_WD2/JRD_CBDMLab/1_JRD_CBDM_PROJECTS/ProjectFolders/T-TregEpigenome/Data/1_Experimental/ATAC/Analyses/A7-all-ttreg/1rpm-cv08/csv/" 
ttrm <- read.csv(file=paste0(path, 'a7-ttreg-1rpm-0.8cv-filtered-rep-avg-counts-log2-rpkm-qn-peakID-mat.csv'),
                 row.names=1, header=T)

# --- create FC matrix ------------------------------------------------------------------------------------------

# use apply() and combn() to compute fold-changes and store as new matrix
fcmat <- apply(combn(ncol(ttrm),2),2,
               function(x) ttrm[,x[1]] / ttrm[,x[2]])
colnames(fcmat) <- apply(combn(ncol(ttrm),2),2,
                         function(x) paste(names(ttrm)[x], collapse="/"))
colnames(fcmat) # check

# for each row of the FC matrix, evaluate whether any value is > FC cutoff, store results in new vectors
FC2_pass <- apply(fcmat,1, function(x) any(x >=2 | x <= (1/2)))
FC3_pass <- apply(fcmat,1, function(x) any(x >=3 | x <= (1/3)))
FC4_pass <- apply(fcmat,1, function(x) any(x >=4 | x <= (1/4)))

# combine into list
fcpl <- list("FC2" = FC2_pass, "FC3"=FC3_pass, "FC4"=FC4_pass) #[f]old[c]hange[p]ass[l]ist

# bind these logical vectors to the original peak matrix
ttrm <- cbind(ttrm, fcpl) 

# create a summary table of number of peaks that pass each FC cutoff
fcst <- data.frame(sapply(fcpl, table))  # [f]old[c]hange[s]ummary[t]able

# --- save FC matrix ----------------------------------------------------------------------------------------------

path <- "/Volumes/JD_WD2/JRD_CBDMLab/1_JRD_CBDM_PROJECTS/ProjectFolders/T-TregEpigenome/Data/1_Experimental/ATAC/Analyses/A7-all-ttreg/1rpm-cv08/csv/" 
n <- 1 # set rpm cutoff for filenames
c <- 0.8 # set cv cutoff for filenames
file <- paste0('a5-ttreg-',n,'rpm-',c,'cv-filtered-fcmat-summary.csv')
write.csv(fcst, paste0(path,file))

# create data frame subsets with peaks passing these criteria
ttrm.fc2.pass <- ttrm[ttrm$FC2 =="TRUE",] # fc >=2 in at least one pair-wise comparison
ttrm.fc2.fail <- ttrm[ttrm$FC2 =="FALSE",] # fc <=2 in all pair-wise comparisons

# remove filter columns 
ttrm.fc2.pass <- ttrm.fc2.pass[,-c(9:11)]
ttrm.fc2.fail <- ttrm.fc2.fail[,-c(9:11)]

# --- save peaks passing FC ----------------------------------------------------------------------------------------

path <- "/Volumes/JD_WD2/JRD_CBDMLab/1_JRD_CBDM_PROJECTS/ProjectFolders/T-TregEpigenome/Data/1_Experimental/ATAC/Analyses/A7-all-ttreg/1rpm-cv08/RDa/"  # set data path
n <- 1 # set rpm cutoff for filenames
c <- 0.8 # set cv cutoff for filenames
fc <- 2 # set fc cutoff for filenames

file <- paste0('a7-ttreg-',n,'rpm-',c,'cv-filtered-FC',fc,'-pass')
save(ttrm.fc2.pass, file = paste0(path,file,'.RDa'))
# load 
load(file = paste(path,file,'.RDa',sep=""))

# --- save peaks failing FC ----------------------------------------------------------------------------------------
path <- "/Volumes/JD_WD2/JRD_CBDMLab/1_JRD_CBDM_PROJECTS/ProjectFolders/T-TregEpigenome/Data/1_Experimental/ATAC/Analyses/A7-all-ttreg/1rpm-cv08/RDa/"  # set data path
n <- 1 # set rpm cutoff for filenames
c <- 0.8 # set cv cutoff for filenames
fc <- 2 # set fc cutoff for filenames

file <- paste0('a7-ttreg-',n,'rpm-',c,'cv-filtered-FC',fc,'-fail')
save(ttrm.fc2.fail, file = paste0(path,file,'.RDa'))
# load 
load(file = paste(path,file,'.RDa',sep=""))

```
<br>  


#### APclust on variable peaks (pair-wise fold-changes of qn-normalized rpkm signal)
```{r apclust_varpeaks}
## ---- load peak samples -----------------------------------------------------------------------------------------
path <- "/Volumes/JD_WD2/JRD_CBDMLab/1_JRD_CBDM_PROJECTS/ProjectFolders/T-TregEpigenome/Data/1_Experimental/ATAC/Analyses/A7-all-ttreg/1rpm-cv08/RDa/"  # set data path
n <- 1 # set rpm cutoff for filenames
c <- 0.8 # set cv cutoff for filenames
fc <- 2 # set fc cutoff for filenames
file <- paste0('a7-ttreg-',n,'rpm-',c,'cv-filtered-FC',fc,'-pass')

load(file = paste(path,file,'.RDa',sep=""))

# convert to matrix
ttrm <- as.matrix(ttrm.fc2.pass)
head(dimnames(ttrm),10)
dim(ttrm)


## ---- APcluster ------------------------------------------------------------------------------------------------
library(apcluster)

# find range of input prefs
#preferenceRange(negDistMat(r=2)(ttrm))
# result = 

# cluster
# params 1
ap1 <- apclusterL(negDistMat(r=2), ttrm, p=-700, frac = 0.5, sweeps = 5)
# 16 clusters

#ap1 <- apclusterL(negDistMat(r=2), ttrm, p=-2000, frac = 0.1, sweeps = 5)

# params 2
ap1 <- apclusterL(negDistMat(r=2), ttrm, p=-3000, frac = 0.1, sweeps = 5)
# 6 clusters 

# params 3
ap1 <- apclusterL(negDistMat(r=2), ttrm, p=-3000, frac = 0.3, sweeps = 5)
# 5 clusters

# params 4 
ap1 <- apclusterL(negDistMat(r=2), ttrm, p=-2000, frac = 0.3, sweeps = 5)
# 8 clusters

# params 5 
ap1 <- apclusterL(negDistMat(r=2), ttrm, p=-2400, frac = 0.3, sweeps = 5)
# 6 clusters
# pursue this one

############ aim for 5 clust?
ap1 <- apclusterL(negDistMat(r=2), ttrm, p=-4000, frac = 0.3, sweeps = 5)



### --- scatter plot and save --------------------------------------------------------------------------------
path <- "/Volumes/JD_WD2/JRD_CBDMLab/1_JRD_CBDM_PROJECTS/ProjectFolders/T-TregEpigenome/Data/1_Experimental/ATAC/Analyses/A7-all-ttreg/1rpm-cv08/plots/"
n <- 1 # set rpm cutoff for filenames
c <- 0.8 # set cv cutoff for filenames
fc <- 2 # set fc cutoff for filenames
k <- 6
file <- paste0('a7-ttreg-',n,'rpm-',c,'cv-filtered-FC',fc,'-pass-apclust-',k)


pdf(file = paste0(path,file,'.pdf'), width=15, height=15)
plot(ap1, ttrm)
dev.off()

png(file = paste0(path,file,'.png'), width=1000, height=1000)
plot(ap1, ttrm)
dev.off()

# eventually plot heatmap but takes long time
# heatmap(ap1)


## proceed with param 5
## ---- append clusterID to peak signal values --------------------------------------------------------------------

# append clustID to peaks
cl <- ap1@clusters # extract AP clusters # [c]luster[l]ist
names(cl) <- c(1:length(cl))

pdfc <- data.frame() #[p]eak[d]ata[f]rame[c]lusters

for (i in 1:length(cl)) { # loop over cluster list and append clusterID to peak
  df <- data.frame(PeakID = names(cl[[i]]), ClusterID = names(cl[i]))
  pdfc <- rbind(pdfc, df)
}

# convert ttrm to dataframe
pdf <- data.frame(PeakID = rownames(ttrm), ttrm, row.names=NULL)  # [p]eak[d]ata[f]rame

# add clusterID onto df
library(dplyr)
pdfc <- left_join(pdfc, pdf, by = 'PeakID') 

# check
summary(pdfc$ClusterID)
# 2171 1706 1486  239  461 1596
ttrm[rownames(ttrm)=='peak_20',]
pdfc[pdfc$PeakID=='peak_20',]

# order
pdfco <- pdfc[order(pdfc$ClusterID),]

# save ordered df
path <- "/Volumes/JD_WD2/JRD_CBDMLab/1_JRD_CBDM_PROJECTS/ProjectFolders/T-TregEpigenome/Data/1_Experimental/ATAC/Analyses/A7-all-ttreg/1rpm-cv08/RDa/"
n <- 1 # set rpm cutoff for filenames
c <- 0.8 # set cv cutoff for filenames
fc <- 2 # set fc cutoff for filenames
k <- 6
file <- paste0('a7-ttreg-',n,'rpm-',c,'cv-filtered-qn-FC',fc,'-pass-apclust-k',k,'pdfco','.RDa')
save(pdfco, file = paste0(path,file))


# to reload
load(paste0(path,file))


##---- pheatmap results across samples ----------------------------------------------------------------------------
# redraw sample/peak heatmap based on cluster results
library(pheatmap)
library(RColorBrewer)

# set colors
niceCols <- colorRampPalette(rev(brewer.pal(11, 'Spectral')))(50)

# get df into shape for heatmap
heat.df <- data.frame(pdfco, row.names = pdfco$PeakID) # make new df for cluster annotations
heat.annot <- heat.df[2]
heat.annot$ClusterID <- as.factor(heat.annot$ClusterID)

# set row gaps based on clusters 
rows <- list()
for (i in 1:6) {
  rows[i] <- nrow(heat.df[heat.df$ClusterID==i,])
}
row_gaps <- unlist(rows)
row_gaps2 <- c(row_gaps[1], sum(row_gaps[1:2]), sum(row_gaps[1:3]), sum(row_gaps[1:4]),
               sum(row_gaps[1:5]))

# draw heatmap based on APclust
pheatmap(as.matrix(heat.df[3:10]), cluster_rows = F, color = niceCols, annotation_row = heat.annot,
         show_rownames = F, border_color = 'black', gaps_row = row_gaps2)


# --- print and save --------------------------------------------------------------------------------------------
path <- "/Volumes/JD_WD2/JRD_CBDMLab/1_JRD_CBDM_PROJECTS/ProjectFolders/T-TregEpigenome/Data/1_Experimental/ATAC/Analyses/A7-all-ttreg/1rpm-cv08/plots/"
n <- 1 # set rpm cutoff for filenames
c <- 0.8 # set cv cutoff for filenames
fc <- 2 # set fc cutoff for filenames
k <- 6
file <- paste0('a7-pheat-qn-',n,'rpm-',c,'cv-','fc',fc,'-pass-apclust-k',k)
# pdf
pdf(paste0(path,file,'.pdf'), width=12, height=12)
pheatmap(as.matrix(heat.df[3:10]), cluster_rows = F, color = niceCols, annotation_row = heat.annot,
         show_rownames = F, border_color = 'black', gaps_row = row_gaps2)
dev.off()

# png
png(paste0(path,file,'.png'), width=1200, height=1200)
pheatmap(as.matrix(heat.df[3:10]), cluster_rows = F, color = niceCols, annotation_row = heat.annot,
         show_rownames = F, border_color = 'black', gaps_row = row_gaps2)
dev.off()


##---- heatmap2 results across samples ----------------------------------------------------------------------------

# not working nicely yet

library(gplots)
library(RColorBrewer)

test <- pdfco
cols <- palette(brewer.pal(8, "Dark2"))[as.numeric(pdfco$ClusterID)]
head(cbind(rownames(test),cols))


# set colors
hmcol <- colorRampPalette(brewer.pal(9, "GnBu"))(100)

# draw
heatmap.2(as.matrix(test[3:10]),
          trace="none", 
          ColSideColors=cols, 
          col=hmcol)

##---- ggplot2 results across samples ----------------------------------------------------------------------------
library(reshape2)
library(ggplot2)
# start with copy of pdfco 
str(pdfco)
x <- pdfco
str(x)
x <- x[c(3:10,1,2)]
str(x)

# sort peaks by cluster
x$peaksort <- x$PeakID[order(x$ClusterID)]
str(x)
x$peaksort <- order(x$ClusterID, rowMeans(x[1:4]))
head(x)

# melt
xm <- melt(x, id.vars=c('PeakID','peaksort'))
xm$value <- as.numeric(xm$value)

# set colors 
niceCols <- colorRampPalette(rev(brewer.pal(11, 'Spectral')))(50)

# plot
ggplot(xm, aes(x=variable, y=peaksort)) + geom_tile(aes(fill=value)) +
scale_fill_gradientn(colors = colorRampPalette(rev(brewer.pal(11, 'Spectral')))(50)) + 
  theme_bw()


## ---- matplots results across samples -------------------------------------------------------------------------
# extract each cluster and combine into a list
cl <- list() # [c]luster [l]ist
for (i in 1:length(unique(pdfco$ClusterID))) {
  cl[[i]] <- as.matrix(pdfco[3:10][pdfco$ClusterID == i,],)
}

# set colors
eqSpect <- colorRampPalette(
               c("#f2003c", "#F66900", "#F19100", "#F1B100",
                 "#EFD300", "#f0ea00", "#CBDB00", "#9DD501",
                 "#5ED108", "#00AF63", "#00A78E", "#00a3ac",
                 "#0093af", "#0082b2", "#006ebf", "#4F37C2",
                 "#8000D3", "#A001BF", "#BA00A4", "#D0007F"),
                 space="rgb",
                 interpolate="spline")
n <- length(unique(pdfco$ClusterID))

niceCols <- eqSpect(n)


# define function to draw matplot for each cluster with mean trendline
matplusmean <- function(x) {
  # x = list of peak matrices
  # draw matplots
  matplot(t(x[[i]]),
          type = 'l', col = niceCols[i],
          xlab = 'Tissue Treg', ylab= 'log2 rpkm',
             xlim=c(1,8),ylim=c(-2,6), xaxt='n')
  # label x axis
  axis(1, at=1:8, labels=c('Cn','Cp','M','V','SCn','SCp','SM','SV'))
  # draw trendline
  points(x = seq(1,8,by=1), y=colMeans(cl[[i]]),lwd=4, col='black',type='l')
}

## ---- draw and save matplots -----------------------------------------------------------------------------------

path <- "/Volumes/JD_WD2/JRD_CBDMLab/1_JRD_CBDM_PROJECTS/ProjectFolders/T-TregEpigenome/Data/1_Experimental/ATAC/Analyses/A7-all-ttreg/1rpm-cv08/plots/"
n <- 1 # set rpm cutoff for filenames
c <- 0.8 # set cv cutoff for filenames
fc <- 2 # set fc cutoff for filenames
k <- 6
file <- paste0('a7-qn-',n,'rpm-',c,'cv-','fc',fc,'-pass-apclust-k',k,'-matplots-mean-trend')

# save and draw
pdf(paste0(path,file,'.pdf'), width=15, height=15)
par(mfrow = c(3,3)) # set rows and columns
# run
for (i in 1:length(unique(pdfco$ClusterID))) {
  matplusmean(cl)
}
# save
dev.off()


# define function to draw matplot for each cluster with median trendline
matplusmeds <- function(x) {
  # x = list of peak matrices
  matplot(t(x[[i]]),
          type = 'l', col = niceCols[i],
          xlab = 'Tissue Treg', ylab= 'log2 rpkm',
             xlim=c(1,8),ylim=c(-2,6),xaxt='n')
  # find median
  med <- apply(x[[i]],2,median)
  # label x axis
  axis(1, at=1:8, labels=c('Cn','Cp','M','V','SCn','SCp','SM','SV'))
  # add trendline
  points(x = seq(1,8,by=1), y=med,lwd=4, col='black',type='l')
}

# set filename
file <- paste0('a7-qn-',n,'rpm-',c,'cv-','fc',fc,'-pass-apclust-k',k,'-matplots-median-trend')

pdf(paste0(path,file,'.pdf'), width=15, height=15)

# set rows and columns
par(mfrow = c(3,3))
# run
for (i in 1:length(unique(pdfco$ClusterID))) {
  matplusmeds(cl)
}
# save
dev.off()


## ---- retrieve peak coordinates and export --------------------------------------------------------------------
# load df with peakID and cluster info
path <- "/Volumes/JD_WD2/JRD_CBDMLab/1_JRD_CBDM_PROJECTS/ProjectFolders/T-TregEpigenome/Data/1_Experimental/ATAC/Analyses/A7-all-ttreg/1rpm-cv08/RDa/"
n <- 1 # set rpm cutoff for filenames
c <- 0.8 # set cv cutoff for filenames
fc <- 2 # set fc cutoff for filenames
k <- 6
file <- paste0('a7-ttreg-',n,'rpm-',c,'cv-filtered-qn-FC',fc,'-pass-apclust-k',k,'pdfco','.RDa')

load(paste0(path,file))

# load df with peakID and chrom coords
path <- "/Volumes/JD_WD2/JRD_CBDMLab/1_JRD_CBDM_PROJECTS/ProjectFolders/T-TregEpigenome/Data/1_Experimental/ATAC/Analyses/A7-all-ttreg/1rpm-cv08/bed/"
filename <- 'a7-ttreg-1rpm-0.8cv-filtered-rep-avg-counts-rpkm-qn.bed'
# read in
ttr.annot <- read.delim(paste0(path,filename), header=T, check.names=T, sep='\t')
# correct colnames
names(ttr.annot) <- c('PeakID','Chr','Start','End','Strand')

# check agreements
pdfco[pdfco$PeakID == 'peak_20',]
ttr.annot[ttr.annot$PeakID == 'peak_20',]

# merge 
library(dplyr)

pdfco.annot <- left_join(pdfco, ttr.annot, by = 'PeakID')
pdfco.annot <- pdfco.annot[c(1,11:14,2:10)]

# save
path <- "/Volumes/JD_WD2/JRD_CBDMLab/1_JRD_CBDM_PROJECTS/ProjectFolders/T-TregEpigenome/Data/1_Experimental/ATAC/Analyses/A7-all-ttreg/1rpm-cv08/RDa/"
n <- 1 # set rpm cutoff for filenames
c <- 0.8 # set cv cutoff for filenames
fc <- 2 # set fc cutoff for filenames
k <- 6
file <- paste0('a7-ttreg-',n,'rpm-',c,'cv-qn-fc',fc,'pass-apk',k,'-peak-df-annot')

save(pdfco.annot, file = paste0(path,file,'.RDa'))

# generate bed format with ClusterID
pdfco.bed <- pdfco.annot[c(2:4,1,6)]

# generate list of peak df split by ClusterID
pl <- list() # [p]eak[l]ist
for (i in 1:length(unique(pdfco.bed$ClusterID))) {
  pl[[i]] <- pdfco.bed[pdfco.bed$ClusterID == i,]
}

# name list elements
pl.names <- paste0('a7-ttreg-',n,'rpm-',c,'cv-fc',fc,'pass-apk',k,'-c')
#pl.names <- 'a5-ttreg-2rpm-fil-qn-fc2-pass-apclust'
pl.names1 <- seq(from = 1,to = length(unique(pdfco.bed$ClusterID)), by = 1)
pl.names2 <- paste(pl.names,pl.names1,sep="")

# name the list elements to match the files
names(pl) <- pl.names2

# write each df to bed file
path <- "/Volumes/JD_WD2/JRD_CBDMLab/1_JRD_CBDM_PROJECTS/ProjectFolders/T-TregEpigenome/Data/1_Experimental/ATAC/Analyses/A7-all-ttreg/1rpm-cv08/bed/"
lapply(names(pl),
       function(x) write.table(pl[[x]][-5], file=paste(paste0(path,x),'bed','txt',sep="."), sep = '\t',
                               quote=F, row.names=F, col.names=c('#Chr','Start','End','PeakID')))
```
<br>  


#### Motif finding in APclusters
+ homer v4.6 run on Orchestra
+ `findMotifsGenome.pl a7-ttreg-1rpm-0.8cv-fc2pass-apk6-c4.bed.txt mm9 a7-apk6-c4-motifs-200/ -size 200 -mask -preparsedDir new/`
```{r cluster_motifs, engine='bash'}
# load module 
module load seq/homer/4.6

# run against whole genome background, -/+200bp, repeat-masked genome
findMotifsGenome.pl a7-ttreg-1rpm-0.8cv-fc2pass-apk6-c4.bed.txt mm9 a7-apk6-c4-motifs-200/ -size 200 -mask -preparsedDir new/
```
<br>  


#### Hclust and Km on all **QN normalized** high-confidence peaks
```{r Hc_Km_allqnpeaks}
## ---- load peaks ---------------------------------------------------------------------------------------------
path <- "/Volumes/JD_WD2/JRD_CBDMLab/1_JRD_CBDM_PROJECTS/ProjectFolders/T-TregEpigenome/Data/1_Experimental/ATAC/Analyses/A7-all-ttreg/1rpm-cv08/csv/" # set data path

ttrm <- read.csv(file=paste0(path, 'a7-ttreg-1rpm-0.8cv-filtered-rep-avg-counts-log2-rpkm-qn-peakID-mat.csv'),
                 row.names=1, header=T)
# convert to matrix
ttrm <- as.matrix(ttrm)
head(dimnames(ttrm),10)

##---- try kmeans function in pheatmap() --------------------------------------------------------------------
library(pheatmap)
library(RColorBrewer)

# set colors
niceCols <- colorRampPalette(rev(brewer.pal(11, 'Spectral')))(50)

set.seed(123456)
# save results of pheatmap to object p
p <- pheatmap(ttrm, color= niceCols, clustering_distance_rows = 'euclidean', clustering_distance_cols = 'euclidean', kmeans_k = 18)

# print
path <- "/Volumes/JD_WD2/JRD_CBDMLab/1_JRD_CBDM_PROJECTS/ProjectFolders/T-TregEpigenome/Data/1_Experimental/ATAC/Analyses/A7-all-ttreg/1rpm-cv08/plots/"
n <- 1 # set rpm cutoff for filenames
c <- 0.8 # set cv cutoff for filenames
k <- 18 # set cluster # for filenames

file <- paste0('a7-ttreg-',n,'rpm-',c,'cv-filtered-qn-allpks-pheat-hc-km-k',k)

# pdf
pdf(paste0(path,file,'.pdf'), width=10, height=10)
p <- pheatmap(ttrm, color= niceCols, clustering_distance_rows = 'minkowski', clustering_distance_cols = 'correlation', kmeans_k = 18)
dev.off()

# png
png(paste0(path,file,'.png'), width=1000,height=1000)
p <- pheatmap(ttrm, color= niceCols, clustering_distance_rows = 'minkowski', clustering_distance_cols = 'correlation', kmeans_k = 18)
dev.off()

## --- extract kmeans results --------------------------------------------------------------------------------------

k <- p$kmeans
# save
path <- "/Volumes/JD_WD2/JRD_CBDMLab/1_JRD_CBDM_PROJECTS/ProjectFolders/T-TregEpigenome/Data/1_Experimental/ATAC/Analyses/A7-all-ttreg/1rpm-cv08/RDa/"
n <- 1 # set rpm cutoff for filenames
c <- 0.8 # set cv cutoff for filenames
kn <- 18 # set cluster # for filenames
save(k, file = paste0(path,'a7-ttreg',n,'rpm-',c,'cv-filtered-qn-allpks-pheat-km-k',kn,'.RDa'))

# extract kmeans clusters
c <- p$kmeans$cluster 
# convert cluster and peak IDs to df
cdf <- data.frame(peak = names(c), cluster = c, row.names= NULL) # [c]luster[d]ata[f]rame
# convert ttreg peak matrix to dataframe
ttr.df <- data.frame(peak = rownames(ttrm), ttrm, row.names=NULL) #[ttr]eg[d]ata[frame]
# add clusterID onto df
library(dplyr)
ttrdfc <- left_join(ttr.df, cdf, by = 'peak') # [ttr]eg[d]ata[f]rame[c]lusters

# check that join worked properly by looking up rpm values in original sample matrix
ttrm[rownames(ttrm)=='peak_5',]
ttrdfc[ttrdfc$peak=='peak_5',]

# save
path <- "/Volumes/JD_WD2/JRD_CBDMLab/1_JRD_CBDM_PROJECTS/ProjectFolders/T-TregEpigenome/Data/1_Experimental/ATAC/Analyses/A7-all-ttreg/1rpm-cv08/RDa/"
n <- 1 # set rpm cutoff for filenames
c <- 0.8 # set cv cutoff for filenames
kn <- 18 # set cluster # for filenames
save(ttrdfc,file=paste0(path,'a7-ttreg-',n,'rpm-',c,'cv-filtered-qn-allpks-pheat-k',kn,'-ttrdfc','.RDa'))

# load
load(paste0(path,'a7-ttreg-1rpm-0.8cv-filtered-qn-allpks-pheat-k18-ttrdfc.RDa'))

## --- order peaks by cluster ID --------------------------------------------------------------------------------

# order df
ttrdfco <- ttrdfc[order(ttrdfc$cluster),] # ttrdfc[o]rdered

# save
path <- "/Volumes/JD_WD2/JRD_CBDMLab/1_JRD_CBDM_PROJECTS/ProjectFolders/T-TregEpigenome/Data/1_Experimental/ATAC/Analyses/A7-all-ttreg/1rpm-cv08/RDa/"
n <- 1 # set rpm cutoff for filenames
c <- 0.8 # set cv cutoff for filenames
kn <- 18 # set cluster # for filenames
save(ttrdfco,file=paste0(path,'a7-ttreg-',n,'rpm-',c,'cv-filtered-qn-allpks-pheat-k',kn,'-ttrdfco','.RDa'))

# redraw heatmap based on kmeans
heat.df <- data.frame(ttrdfco, row.names = ttrdfco$peak) # make new df for cluster annotations 

heat.annot <- heat.df[-c(1:9)]
heat.annot$cluster <- as.factor(heat.annot$cluster)
pheatmap(as.matrix(heat.df[2:9]), cluster_rows = F, color = niceCols, annotation_row = heat.annot,
         show_rownames = F)

# print and save pdf
path <- "/Volumes/JD_WD2/JRD_CBDMLab/1_JRD_CBDM_PROJECTS/ProjectFolders/T-TregEpigenome/Data/1_Experimental/ATAC/Analyses/A7-all-ttreg/1rpm-cv08/plots/"
n <- 1 # set rpm cutoff for filenames
c <- 0.8 # set cv cutoff for filenames
kn <- 18 # set cluster # for filenames
file <- paste0('a7-ttreg-',n,'rpm-',c,'cv-filtered-qn-allpks-pheat-km-k',kn,'-allrows','.pdf')
pdf(paste0(path,file))
pheatmap(as.matrix(heat.df[2:9]), cluster_rows = F, color = niceCols, annotation_row = heat.annot,
         show_rownames = F)
dev.off()

# print and save png
file <- paste0('a7-ttreg-',n,'rpm-',c,'cv-filtered-qn-allpks-pheat-km-k',kn,'-allrows','.png')
png(paste0(path,file), width=1200, height=1200)
pheatmap(as.matrix(heat.df[2:9]), cluster_rows = F, color = niceCols, annotation_row = heat.annot,
         show_rownames = F)
dev.off()
```
<br>  


#### HOLD: Clustering by hand
+ start with log2 rpkm matrix of replicate averages
+ # start with 1.5 (log2 qn rpkm)
+ iterate following steps over each t-Treg population:
+ what fraction of total t-Treg peaks are present?
+ of those, how many are shared with 0,1,2,3,4,5,6,7,or 8 other t-Treg populations
+ then, for each of these "clusters", perform unsupervised hc/Km/other to find sub-patterns
+ so far this doesn't work bc diffs between populations are not absolute but relative
```{r clust_by_hand}
## ----- load count data -----------------------------------------------------------------------------------------

path <- "/Volumes/JD_WD2/JRD_CBDMLab/1_JRD_CBDM_PROJECTS/ProjectFolders/T-TregEpigenome/Data/1_Experimental/ATAC/Analyses/A7-all-ttreg/1rpm-cv08/RDa/"
filename <- 'a7-ttreg-1rpm-0.8cv-filtered-rep-avg-counts-log2-rpkm-qn.RDa'
load(paste0(path,filename))

## -----look at distributions -------------------------------------------------------------------------------------
str(ttr.avg.l2)
summary(ttr.avg.l2[6:13])

# boxplot
par(oma = c(4,1,0.5,1))
boxplot(ttr.avg.l2[6:13], las=2, cex.axis=0.7, ylab='log2 rpkm',
        main=paste('merged atac peaks','n=',nrow(ttr.avg.l2)))

# histograms
library(reshape2)
library(ggplot2)
# get data into shape for ggplot
ttr.melt <- melt(ttr.avg.l2, id.vars=c('CHR','START','END','PeakID','Strand'), variable.name = 'SampleID',
                 value.name = 'log2_rpkm')
head(ttr.melt)
tail(ttr.melt)

g <- ggplot(data = ttr.melt, aes(x=log2_rpkm))
g + geom_histogram(fill="white",color="black",binwidth=0.1) +
  facet_grid(SampleID~.) +
  scale_x_continuous(limits=c(-4, 12), breaks=seq(-4,12,by=1)) +
  geom_vline(xintercept =log(1,2)) +
   geom_vline(xintercept = log(2,2)) +
  geom_vline(xintercept = log(3,2)) +
  theme_bw()

# need to figure out threshold based on genome browser
# start with 1.5 (log2 qn rpkm)

# Cln_Nrp1n
test <- ttr.avg.l2

# filter for peaks present in Cln_Nrp1n
library(dplyr)
Cln_Nrp1n <- filter(ttr.avg.l2, Cln_Nrp1n > 1.5)
uniq <- apply(Cln_Nrp1n[7:13], 1, function(x) all(x > 1.5))
summary(uniq)



Vat <- filter(ttr.avg.l2, Vat >= 3)
Vat.uniq <- apply(Vat[,c(6:8,10:13)], 1, function(x) all(x <= 2.5))
summary(Vat.uniq)
VAT.test <- Vat[Vat.uniq,]


Colon <- filter(ttr.avg.l2, Cln_Nrp1n > 3 & Cln_Nrp1p > 3)
Colon.uniq <- apply(Colon[8:13], 1, function(x) all(x <= 2))
summary(Colon.uniq)
test <- Colon[Colon.uniq,]



Colon.uniq <- lapply(Colon[,8:13], function(x) x <= 1.5)
lapply(Colon.uniq, summary)
summary(Colon.uniq)

FC2_logical <- apply(ATAC_6way_FCmatrix,1, function(x) any(x >=2 | x <= (1/2)))


```
<br>    





