---
title: "A8-ATAC-allsamples"
author: "JRD"
date: "March 28, 2016"
output: html_document
---

last modified: `r Sys.Date()`
<br>    

#### README
+ this document contains code and explanation for an analysis comparing Colon, Muscle, Splenic and VAT Treg
+ for details on how the input files used in this analysis were generated, see *ATAC_mapping.Rmd* and *ATAC_peaks.Rmd*
+ **difference between A5 and A7 is latter filters peaks by both min signal and replicate CV**
<br>    


#### R Markdown Specifications
Set global options
```{r setoptions, echo=F}
library(knitr)
opts_chunk$set(echo=F, eval=F, warning=F, message=F)
```
<br>    

Github
https://github.com/jdis24/ATAC.git
<br> 


#### Questions answered by plots
+ are there differences between results of homer factor and dnase peak-calling modes?
+ how well do identification and signal distribution of peaks correlate across biological replicates?


#### Pseudo
**call peaks**
+ [code for these steps is in "ATAC_peaks.Rmd"]
+ for each sample, call peaks in each replicate (homer factor and dnase modes)
+ copied peak calling results from Orchestra ATAC folders for each batch to folder for this analysis
+ [code for all subsequent steps is in this document]
**compare results of dnase and factor modes for each replicate**
+ extract these vectors from list of filenames: Tissue, CellType, SampleName, SampleID, Replicate
+ loop over each output file and extract summary stats: total peaks, total reads, frag length, reads in peaks, rip%, fdr threshold reads
+ loop over each output file and extract peak stats: mean / median / stdev / var of width,score
+ add to summary table
+ plots


**merge results of dnase and factor modes for each replicate**
+ homer, with code copied here

**QC, get read counts, filter**
+ merge results  of dnase and factor modes into single peakset for each replicate (homer)
+ merge replicate peaksets,keeping all, but retain whether peak was called in either or both (homer)
+ for each replicate, count reads in merged peakset, normalized to total library size (ChIPQC)
+ scatterplot replicates with r2
+ filter for peaks called in both replicates (using homer or using min rpkm?)
+ export .csv files of peaks and counts


#### Desired plots:
pre-replicate merge
master table: cell_number, total reads, uniquely aligned, duplicated, %M, final, avg frag length, # peaks, % RIP, 
table w/ # reads in peaks (csv, xtable?)
scatterplot total # vs reads in peaks
table w/ # peaks at different thresholds (csv, xtable?)
boxplot variation in peak # at different thresholds
final table of peaks at desired threshold
**colors** 
brewer.pal(10, "Spectral")
incorporate sample labels 
<br> 


#### Compare results of dnase and factor modes for each replicate
```{r dnase_vs_factor}
# path to peak files
path <- "/Volumes/JD_WD2/JRD_CBDMLab/1_JRD_CBDM_PROJECTS/ProjectFolders/T-TregEpigenome/Data/1_Experimental/ATAC/Analyses/A8-all-samples/peaks/dnaseMode/"

#### -- extract sample information from filenames ----------------------------------------------------------------
# get filenames to use for extracting sample info
fn <- list.files(path = path, pattern = ".txt")

# tissue
tissue1 <- lapply(fn, strsplit, split="_")
tissue <- unlist(lapply(tissue1, function(x) x[[1]][[1]]))

# celltype
ct1 <- lapply(fn, strsplit, split="_rep*")
ct2 <- lapply(ct1, function(x) x[[1]][[1]])
ct3 <- list()
for (i in 1:length(ct2)) {
  if (grepl("Treg",ct2[[i]])) {
    ct3[i]<- substr(ct2[[i]], start=nchar(ct2[[i]])-3,stop=nchar(ct2[[i]]))
  } else
    if (grepl("Tconv",ct2[[i]])) {
      ct3[i] <- substr(ct2[[i]], start=nchar(ct2[[i]])-4,stop=nchar(ct2[[i]]))
    }
    else
    ct3[i]<- substr(ct2[[i]], start=nchar(ct2[[i]])-2, stop=nchar(ct2[[i]]))
}
celltype <- unlist(ct3)

# sample name
samplename1 <- unlist(lapply(fn, strsplit, ".peaks*"))
samplename <- samplename1[grep(".dnase.txt",samplename1, invert=T)]

# sample ID
sampleid1 <- sapply(fn, strsplit, ".rep")
sampleid <- unlist(lapply(sampleid1, function(x) x[[1]]))
sampleid <- unname(sampleid)

# replicate
reps1 <- sapply(fn, strsplit, "_rep*")
reps2 <- sapply(reps1, function(x) strsplit(x[[2]], ".peaks"))
reps3 <- as.character(sapply(reps2, function(x) x[[1]]))
reps4 <- gsub("1.2","1",reps3)
reps <- gsub("3","2",reps4)


#### -- extract summary info from dnase peaks --------------------------------------------------------------------
library(gsubfn)
library(dplyr)

# store filenames with extensions
fn <- list.files(path = path, pattern = ".txt", full.names=T)

# also read into list
fl <- list()
fl <- lapply(fn, read.table)

# function to read peak totals and parameters from homer peak output file header
extract_peak_info <- function(file) {
  # read in lines based on pattern matching and  
  # extract numeric value following strings
  total_peaks <- as.numeric(read.pattern(file = file, text = file, pattern = "total peaks = *(\\d+)"))
  
  total_reads <- as.numeric(read.pattern(file = file, text = file, pattern = "Total tags = *(\\d+)"))
  
  fragment_length <- as.numeric(read.pattern(file = file, text = file, pattern = "fragment length = *(\\d+)"))
  
  reads_in_peaks <- as.numeric(read.pattern(file = file, text = file, pattern = "Total tags in peaks = *(\\d+)"))
  
  rip_pct <- as.numeric(read.pattern(file = file, text = file, pattern = "Approximate IP efficiency = *(\\d+.\\d+)"))
  
  fdr_read_threshold <- as.numeric(read.pattern(file = file, text = file, pattern = "FDR tag threshold = *(\\d+)"))
  
  output <- c(total_peaks = total_peaks, 
                 total_reads = total_reads,
                 fragment_length = fragment_length,
                 reads_in_peaks = reads_in_peaks,
                 rip_pct = rip_pct,
                 fdr_read_threshold = fdr_read_threshold)
  return(output)
}


#### -- extract peak stats from dnase peaks --------------------------------------------------------------------

# set column names
headers1 <- read.table(fn[[1]], skip=33, nrow=1, comment.char="", sep="\t")
headers <- as.character(unlist(headers1))

for (i in 1:length(fl)) {
  names(fl[[i]]) <- headers
}

# function to calculate peak summary stats from homer peak table
extract_peak_stats <- function(df) {
  df2 <- df %>%
    mutate_(PeakWidth = df$end - df$start)
  
 output <- df2 %>% 
    summarise(PeakWidthMean = mean(PeakWidth),
              PeakWidthMedian = median(PeakWidth),
              PeakWidthSD = sd(PeakWidth),
              PeakWidthVar = var(PeakWidth),
              PeakRpmMean = mean(`Normalized Tag Count`),
              PeakRpmMedian = median(`Normalized Tag Count`),
              PeakRpmSD = sd(`Normalized Tag Count`),
              PeakRpmVar = var(`Normalized Tag Count`))
}

# sapply the functions over the vector of peak filenames
peakStats1 <- t(sapply(fn, extract_peak_info, USE.NAMES = F))
peakStats2 <- t(sapply(fl, extract_peak_stats, USE.NAMES = F))

# combine into summary table 
mode <- "dnase"
table1 <- data.frame(Tissue = tissue, 
                    CellType = celltype, 
                    SampleName = samplename, 
                    SampleID = sampleid,
                    Replicate = reps, 
                    Mode = mode,
                    peakStats1,
                    peakStats2
                    )

#### -- repeat for factor mode peaks ------------------------------------------------------------------------------
# change path
path <- "/Volumes/JD_WD2/JRD_CBDMLab/1_JRD_CBDM_PROJECTS/ProjectFolders/T-TregEpigenome/Data/1_Experimental/ATAC/Analyses/A8-all-samples/peaks/factorMode/"

# store filenames with extensions
fn <- list.files(path = path, pattern = ".txt", full.names=T)

# also read into list
fl <- list()
fl <- lapply(fn, read.table)

# set column names
headers1 <- read.table(fn[[1]], skip=34, nrow=1, comment.char="", sep="\t")
headers <- as.character(unlist(headers1))

for (i in 1:length(fl)) {
  names(fl[[i]]) <- headers
}

# sapply the function over the vector of peak filenames
peakStats3 <- t(sapply(fn, extract_peak_info, USE.NAMES = F))
peakStats4 <- t(sapply(fl, extract_peak_stats, USE.NAMES = F))


#### -- combine into summary table ------------------------------------------------------------------------------
mode <- "factor"
table2 <- data.frame(Tissue = tissue, 
                    CellType = celltype, 
                    SampleName = samplename, 
                    SampleID = sampleid,
                    Replicate = reps, 
                    Mode = mode,
                    peakStats3,
                    peakStats4
                    )


#### -- combine results of dnase and factor modes into single table --------------------------------------------
peak_stat_table <- rbind(table1,table2)

# convert numeric columns, which are lists, into atomic vectors
peak_stat_table[13:20] <- data.frame(lapply(peak_stat_table[13:20], as.numeric))


#### -- export -------------------------------------------------------------------------------------------------
path <- "/Volumes/JD_WD2/JRD_CBDMLab/1_JRD_CBDM_PROJECTS/ProjectFolders/T-TregEpigenome/Data/1_Experimental/ATAC/Analyses/A8-all-samples/peaks/"
save(peak_stat_table, file = paste0(path,"peak_stat_table",".RDa"))
# fix write.csv
write.csv(test, file = paste0(path,"peak_stat_table",".csv"), row.names = F)
```
<br>


#### Summary plots
```{r dnase_factor_plots}
library(dplyr)
library(tidyr)
library(reshape2)
library(scales)
library(ggplot2)
library(RColorBrewer)

#### -- reshape data table ----------------------------------------------------------------------------------
# copy peak table
dat <- peak_stat_table

# convert read columns to K and M
dat$total_peaks <- dat$total_peaks/1000
dat$total_reads <- dat$total_reads/1000000
dat$reads_in_peaks <- dat$reads_in_peaks/1000000

# extract dnase stats to table1
t1 <- dat[dat$Mode=="dnase",]
# melt
t1melted <- melt(t1, id.vars = c(names(t1[1:6])), value.name = "d_value")

# extract factor stats to table2
t2 <- dat[dat$Mode=="factor",]
# melt
t2melted <- melt(t2, id.vars = c(names(t2[1:6])), value.name = "f_value")

# remove mode from t1
t1melted <- subset(t1melted, select = -Mode)
# append factor values
t1melted$f_value <- t2melted$f_value

# rename
dat2 <- t1melted


#### -- plot multiple stats for dnase vs factor modes -----------------------------------------------------------
g <- ggplot(dat2, aes(x=d_value, y=f_value, color = SampleID)) 

g +
  geom_point() + 
  facet_wrap(~variable, scales = c('free')) + 
  scale_x_continuous(breaks=pretty_breaks()) + 
  scale_y_continuous(breaks=pretty_breaks()) + 
  xlab("DNAse Mode") + 
  ylab("Factor Mode") +
  scale_color_brewer(palette = "Spectral") +
  theme_bw() +
  theme(legend.position = "top") 
 
# save
path <- "/Volumes/JD_WD2/JRD_CBDMLab/1_JRD_CBDM_PROJECTS/ProjectFolders/T-TregEpigenome/Data/1_Experimental/ATAC/Analyses/A8-all-samples/plots/"
ggsave(file = paste0(path,'dnase_vs_factor','.pdf'), width = 14, height = 12)
```



#### Merge results of dnase and factor modes for each replicate
+ Orchestra, homer 4.6
```{r merge_dnase_factor, engine = "bash"}
# load module 

```
<br>



