---
title: "A8-ATAC-allsamples"
author: "JRD"
date: "March 28, 2016"
output: html_document
---

last modified: `r Sys.Date()`
<br>    

#### README
+ this document contains code and explanation for an analysis comparing Colon, Muscle, Splenic and VAT Treg
+ for details on how the input files used in this analysis were generated, see *ATAC_mapping.Rmd* and *ATAC_peaks.Rmd*
+ **difference between A5 and A7 is latter filters peaks by both min signal and replicate CV**
<br>    


#### R Markdown Specifications
Set global options
```{r setoptions, echo=F}
library(knitr)
opts_chunk$set(echo=F, eval=F, warning=F, message=F)
```
<br>    

Github
https://github.com/jdis24/ATAC.git
<br> 


#### Questions answered by plots
+ are there differences between results of homer factor and dnase peak-calling modes?
+ how well do identification and signal distribution of peaks correlate across biological replicates?


#### Pseudo
**call peaks**
+ [code for these steps is in "ATAC_peaks.Rmd"]
+ for each sample, call peaks in each replicate (homer factor and dnase modes)
+ copied peak calling results from Orchestra ATAC folders for each batch to folder for this analysis
+ [code for all subsequent steps is in this document]
**compare results of dnase and factor modes for each replicate**
+ extract these vectors from list of filenames: Tissue, CellType, SampleName, SampleID, Replicate
+ loop over each output file and extract summary stats: total peaks, total reads, frag length, reads in peaks, rip%, fdr threshold reads
+ loop over each output file and extract peak stats: mean / median / stdev / var of width,score
+ add to summary table
+ plots


**merge results of dnase and factor modes for each replicate**
+ homer, with code copied here

**QC, get read counts, filter**
+ merge results  of dnase and factor modes into single peakset for each replicate (homer)
+ merge replicate peaksets,keeping all, but retain whether peak was called in either or both (homer)
+ for each replicate, count reads in merged peakset, normalized to total library size (ChIPQC)
+ scatterplot replicates with r2
+ filter for peaks called in both replicates (using homer or using min rpkm?)
+ export .csv files of peaks and counts


#### Desired plots:
pre-replicate merge
master table: cell_number, total reads, uniquely aligned, duplicated, %M, final, avg frag length, # peaks, % RIP, 
table w/ # reads in peaks (csv, xtable?)
scatterplot total # vs reads in peaks
table w/ # peaks at different thresholds (csv, xtable?)
boxplot variation in peak # at different thresholds
final table of peaks at desired threshold
**colors** 
brewer.pal(10, "Spectral")
incorporate sample labels 
<br> 


#### Compare results of dnase and factor modes for each replicate
```{r dnase_vs_factor}
# path to peak files
path <- "/Volumes/JD_WD2/JRD_CBDMLab/1_JRD_CBDM_PROJECTS/ProjectFolders/T-TregEpigenome/Data/1_Experimental/ATAC/Analyses/A8-all-samples/peaks/dnaseMode/"

#### -- extract sample information from filenames ----------------------------------------------------------------
# get filenames to use for extracting sample info
fn <- list.files(path = path, pattern = ".txt")

# tissue
tissue1 <- lapply(fn, strsplit, split="_")
tissue <- unlist(lapply(tissue1, function(x) x[[1]][[1]]))

# celltype
ct1 <- lapply(fn, strsplit, split="_rep*")
ct2 <- lapply(ct1, function(x) x[[1]][[1]])
ct3 <- list()
for (i in 1:length(ct2)) {
  if (grepl("Treg",ct2[[i]])) {
    ct3[i]<- substr(ct2[[i]], start=nchar(ct2[[i]])-3,stop=nchar(ct2[[i]]))
  } else
    if (grepl("Tconv",ct2[[i]])) {
      ct3[i] <- substr(ct2[[i]], start=nchar(ct2[[i]])-4,stop=nchar(ct2[[i]]))
    }
    else
    ct3[i]<- substr(ct2[[i]], start=nchar(ct2[[i]])-2, stop=nchar(ct2[[i]]))
}
celltype <- unlist(ct3)

# sample name
samplename1 <- unlist(lapply(fn, strsplit, ".peaks*"))
samplename <- samplename1[grep(".dnase.txt",samplename1, invert=T)]

# sample ID
sampleid1 <- sapply(fn, strsplit, ".rep")
sampleid <- unlist(lapply(sampleid1, function(x) x[[1]]))
sampleid <- unname(sampleid)

# replicate
reps1 <- sapply(fn, strsplit, "_rep*")
reps2 <- sapply(reps1, function(x) strsplit(x[[2]], ".peaks"))
reps3 <- as.character(sapply(reps2, function(x) x[[1]]))
reps4 <- gsub("1.2","1",reps3)
reps <- gsub("3","2",reps4)


#### -- extract summary info from dnase peaks --------------------------------------------------------------------
library(gsubfn)
library(dplyr)

# store filenames with extensions
fn <- list.files(path = path, pattern = ".txt", full.names=T)

# also read into list
fl <- list()
fl <- lapply(fn, read.table)

# function to read peak totals and parameters from homer peak output file header
extract_peak_info <- function(file) {
  # read in lines based on pattern matching and  
  # extract numeric value following strings
  total_peaks <- as.numeric(read.pattern(file = file, text = file, pattern = "total peaks = *(\\d+)"))
  
  total_reads <- as.numeric(read.pattern(file = file, text = file, pattern = "Total tags = *(\\d+)"))
  
  fragment_length <- as.numeric(read.pattern(file = file, text = file, pattern = "fragment length = *(\\d+)"))
  
  reads_in_peaks <- as.numeric(read.pattern(file = file, text = file, pattern = "Total tags in peaks = *(\\d+)"))
  
  rip_pct <- as.numeric(read.pattern(file = file, text = file, pattern = "Approximate IP efficiency = *(\\d+.\\d+)"))
  
  fdr_read_threshold <- as.numeric(read.pattern(file = file, text = file, pattern = "FDR tag threshold = *(\\d+)"))
  
  output <- c(total_peaks = total_peaks, 
                 total_reads = total_reads,
                 fragment_length = fragment_length,
                 reads_in_peaks = reads_in_peaks,
                 rip_pct = rip_pct,
                 fdr_read_threshold = fdr_read_threshold)
  return(output)
}


#### -- extract peak stats from dnase peaks --------------------------------------------------------------------

# set column names
headers1 <- read.table(fn[[1]], skip=33, nrow=1, comment.char="", sep="\t")
headers <- as.character(unlist(headers1))

for (i in 1:length(fl)) {
  names(fl[[i]]) <- headers
}

# function to calculate peak summary stats from homer peak table
extract_peak_stats <- function(df) {
  df2 <- df %>%
    mutate_(PeakWidth = df$end - df$start)
  
 output <- df2 %>% 
    summarise(PeakWidthMean = mean(PeakWidth),
              PeakWidthMedian = median(PeakWidth),
              PeakWidthSD = sd(PeakWidth),
              PeakWidthVar = var(PeakWidth),
              PeakRpmMean = mean(`Normalized Tag Count`),
              PeakRpmMedian = median(`Normalized Tag Count`),
              PeakRpmSD = sd(`Normalized Tag Count`),
              PeakRpmVar = var(`Normalized Tag Count`))
}

# sapply the functions over the vector of peak filenames
peakStats1 <- t(sapply(fn, extract_peak_info, USE.NAMES = F))
peakStats2 <- t(sapply(fl, extract_peak_stats, USE.NAMES = F))

# combine into summary table 
mode <- "dnase"
table1 <- data.frame(Tissue = tissue, 
                    CellType = celltype, 
                    SampleName = samplename, 
                    SampleID = sampleid,
                    Replicate = reps, 
                    Mode = mode,
                    peakStats1,
                    peakStats2
                    )

#### -- repeat for factor mode peaks ------------------------------------------------------------------------------
# change path
path <- "/Volumes/JD_WD2/JRD_CBDMLab/1_JRD_CBDM_PROJECTS/ProjectFolders/T-TregEpigenome/Data/1_Experimental/ATAC/Analyses/A8-all-samples/peaks/factorMode/"

# store filenames with extensions
fn <- list.files(path = path, pattern = ".txt", full.names=T)

# also read into list
fl <- list()
fl <- lapply(fn, read.table)

# set column names
headers1 <- read.table(fn[[1]], skip=34, nrow=1, comment.char="", sep="\t")
headers <- as.character(unlist(headers1))

for (i in 1:length(fl)) {
  names(fl[[i]]) <- headers
}

# sapply the function over the vector of peak filenames
peakStats3 <- t(sapply(fn, extract_peak_info, USE.NAMES = F))
peakStats4 <- t(sapply(fl, extract_peak_stats, USE.NAMES = F))


#### -- combine into summary table ------------------------------------------------------------------------------
mode <- "factor"
table2 <- data.frame(Tissue = tissue, 
                    CellType = celltype, 
                    SampleName = samplename, 
                    SampleID = sampleid,
                    Replicate = reps, 
                    Mode = mode,
                    peakStats3,
                    peakStats4
                    )


#### -- combine results of dnase and factor modes into single table --------------------------------------------
peak_stat_table <- rbind(table1,table2)

# convert numeric columns, which are lists, into atomic vectors
peak_stat_table[13:20] <- data.frame(lapply(peak_stat_table[13:20], as.numeric))


#### -- export -------------------------------------------------------------------------------------------------
path <- "/Volumes/JD_WD2/JRD_CBDMLab/1_JRD_CBDM_PROJECTS/ProjectFolders/T-TregEpigenome/Data/1_Experimental/ATAC/Analyses/A8-all-samples/peaks/"
save(peak_stat_table, file = paste0(path,"peak_stat_table",".RDa"))
# fix write.csv
write.csv(test, file = paste0(path,"peak_stat_table",".csv"), row.names = F)
```
<br>


#### Summary plots
```{r dnase_factor_plots}
library(dplyr)
library(tidyr)
library(reshape2)
library(scales)
library(ggplot2)
library(RColorBrewer)
#### -- load peak stat table --------------------------------------------------------------------------------
# import peak stat table
path <- "/Volumes/JD_WD2/JRD_CBDMLab/1_JRD_CBDM_PROJECTS/ProjectFolders/T-TregEpigenome/Data/1_Experimental/ATAC/Analyses/A8-all-samples/peaks/"
file <- "peak_stat_table.RDa"
load(paste0(path,file))


#### -- reshape data table ----------------------------------------------------------------------------------
# copy peak table
dat <- peak_stat_table

# convert read columns to K and M
dat$total_peaks <- dat$total_peaks/1000
dat$total_reads <- dat$total_reads/1000000
dat$reads_in_peaks <- dat$reads_in_peaks/1000000

# extract dnase stats to table1
t1 <- dat[dat$Mode=="dnase",]
# melt
t1melted <- melt(t1, id.vars = c(names(t1[1:6])), value.name = "d_value")

# extract factor stats to table2
t2 <- dat[dat$Mode=="factor",]
# melt
t2melted <- melt(t2, id.vars = c(names(t2[1:6])), value.name = "f_value")

# remove mode from t1
t1melted <- subset(t1melted, select = -Mode)
# append factor values
t1melted$f_value <- t2melted$f_value

# rename
peak_stat_table_melted <- t1melted


#### -- plot multiple stats for dnase vs factor modes -----------------------------------------------------------
g <- ggplot(peak_stat_table_melted, aes(x=d_value, y=f_value, color = SampleID)) 

g +
  geom_point() + 
  facet_wrap(~variable, scales = c('free')) + 
  scale_x_continuous(breaks=pretty_breaks()) + 
  scale_y_continuous(breaks=pretty_breaks()) + 
  xlab("DNAse Mode") + 
  ylab("Factor Mode") +
  scale_color_brewer(palette = "Spectral") +
  theme_bw() +
  theme(legend.position = "top") 
 
# save
path <- "/Volumes/JD_WD2/JRD_CBDMLab/1_JRD_CBDM_PROJECTS/ProjectFolders/T-TregEpigenome/Data/1_Experimental/ATAC/Analyses/A8-all-samples/plots/"
ggsave(file = paste0(path,'dnase_vs_factor','.pdf'), width = 14, height = 12)
```
<br>


#### Merge results of dnase and factor modes for each replicate
+ Orchestra, homer 4.6
```{r merge_dnase_factor, engine = "bash"}
# load module 
module load seq/homer/4.6

# run, write stats to separate file
mergePeaks Vat_Treg_rep2.peaks.dnase.txt Vat_Treg_rep2.peaks.factor.txt -venn Vat_Treg_rep2.peaks.homer.venn.txt > Vat_Treg_rep2.peaks.homer.txt
```
<br>


#### Stats for mode merge
```{r merge_dnase_factor_stats}
#### -- load summary stat files from merge ------------------------------------------------------------------------
path <- "/Volumes/JD_WD2/JRD_CBDMLab/1_JRD_CBDM_PROJECTS/ProjectFolders/T-TregEpigenome/Data/1_Experimental/ATAC/Analyses/A8-all-samples/peaks/dnase-factor-merged-100bp-venn/"


#### -- extract sample information from filenames ----------------------------------------------------------------
# read file names
fn <- list.files(path = path, pattern = ".txt")

# extract sample names
samplename <- unlist(lapply(fn, strsplit, ".peaks.homer.venn.txt"))

#### -- extract total number peaks after merge ------------------------------------------------------------------
path <- "/Volumes/JD_WD2/JRD_CBDMLab/1_JRD_CBDM_PROJECTS/ProjectFolders/T-TregEpigenome/Data/1_Experimental/ATAC/Analyses/A8-all-samples/peaks/dnase-factor-merged-100bp/"

# store filenames with extensions
fn <- list.files(path = path, pattern = ".txt", full.names=T)

# also read into list
fl <- list()
fl <- lapply(fn, read.table, sep="\t")

# extract total number peaks 
merged_peak_totals <- sapply(fl, nrow)


#### -- extract peak numbers from files --------------------------------------------------------------------------
library(gsubfn)
library(dplyr)

# set path
path <- "/Volumes/JD_WD2/JRD_CBDMLab/1_JRD_CBDM_PROJECTS/ProjectFolders/T-TregEpigenome/Data/1_Experimental/ATAC/Analyses/A8-all-samples/peaks/dnase-factor-merged-100bp-venn/"

# store filenames with extensions
fn <- list.files(path = path, pattern = ".txt", full.names=T)

# also read into list
fl <- list()
fl <- lapply(fn, read.table, skip=1, sep="\t")

# store column names
headers <- c("factor_only", "dnase_only","common","File")

# name columns within list elements
for (i in 1:length(fl)) {
  names(fl[[i]]) <- headers
}

# extract peak numbers from each df in list
totals <- as.data.frame(t(sapply(fl, function(x) unlist(x$common))))

# name columns
names(totals) <- c("factor_only", "dnase_only","common")

# compute totals
totals$factor_total <- totals$factor_only + totals$common
totals$dnase_total <- totals$dnase_only + totals$common
totals$merge_total <- merged_peak_totals
totals$pct_concordant <- (totals$common / totals$merge_total)*100
totals$pct_dnase_only <- (totals$dnase_only / totals$merge_total)*100
totals$pct_factor_only <- (totals$factor_only / totals$merge_total)*100


#### -- merge into table ---------------------------------------------------------------------------------------
merge_result_table <- data.frame(SampleName =  samplename,
                                 totals)
                                 

#### -- export ---------------------------------------------------------------------------------------------------
path <- "/Volumes/JD_WD2/JRD_CBDMLab/1_JRD_CBDM_PROJECTS/ProjectFolders/T-TregEpigenome/Data/1_Experimental/ATAC/Analyses/A8-all-samples/peaks/"
save(merge_result_table, file = paste0(path,"mode_merge_result_table",".RDa"))

# csv
write.csv(merge_result_table, file = paste0(path,"mode_merge_result_table",".csv"))
```
<br>


#### Plot stats for mode merge
```{r merge_dnase_factor_plots}
library(reshape2)
library(RColorBrewer)
library(scales)
library(ggplot2)
#### -- load summary stat files from merge ------------------------------------------------------------------------
path <- "/Volumes/JD_WD2/JRD_CBDMLab/1_JRD_CBDM_PROJECTS/ProjectFolders/T-TregEpigenome/Data/1_Experimental/ATAC/Analyses/A8-all-samples/peaks/"
load(paste0(path,"merge_result_table,",".RDa"))


#### -- plot stacked bar graph -----------------------------------------------------------------------------------
# copy df
dat <- merge_result_table

# melt
datm <- melt(dat, id.vars = c(names(dat[1:7])))

# plot
g <- ggplot(datm, aes(x=SampleName, y= value, fill=variable))
g +
  geom_bar(stat="identity", color="black") + 
  guides(fill=guide_legend(reverse=T)) +
  scale_fill_brewer(palette="Spectral") +
  coord_flip() +
  scale_x_discrete(limits=rev(levels(datm$SampleName))) + 
  ylab("Percent of Total") +
  theme_bw() 

# save
path <- "/Volumes/JD_WD2/JRD_CBDMLab/1_JRD_CBDM_PROJECTS/ProjectFolders/T-TregEpigenome/Data/1_Experimental/ATAC/Analyses/A8-all-samples/plots/"
ggsave(paste0(path,"dnase_factor_pct_total.pdf"), width=8, height = 7)
```
<br>  


#### Compare sample replicates by merging peaks 
+ Orchestra, homer 4.6
+ separate analyses, changing max peak distance to merge
```{r }
# load module 
module load seq/homer/4.6

# run with default settings, write stats to separate file
mergePeaks Mus_Treg_rep1.peaks.homer.txt Mus_Treg_rep2.peaks.homer.txt -venn Mus_Treg.venn.homer.txt > Mus_Treg.homer.txt

# run, changing max peak center distance to merge = 50, rather than default = 100
mergePeaks Mus_Treg_rep1.peaks.homer.txt Mus_Treg_rep2.peaks.homer.txt -d 50 -venn Mus_Treg.50bp.venn.homer.txt > Mus_Treg.50bp.homer.txt

# run, changing max peak center distance to merge = 0, rather than default = 100
mergePeaks Mus_Treg_rep1.peaks.homer.txt Mus_Treg_rep2.peaks.homer.txt -d 0 -venn Mus_Treg.0bp.venn.homer.txt > Mus_Treg.0bp.homer.txt

```
<br> 


#### Stats from merging replicates
+ extract from homer output files and generate tables 
```{r rep_merge_stats}
#### -- extract sample information from filenames ----------------------------------------------------------------
# set path
path <- "/Volumes/JD_WD2/JRD_CBDMLab/1_JRD_CBDM_PROJECTS/ProjectFolders/T-TregEpigenome/Data/1_Experimental/ATAC/Analyses/A8-all-samples/peaks/replicates-merged-100bp-venn/"

# read file names
fn <- list.files(path = path, pattern = ".txt")

# extract sample names
samplename <- unlist(lapply(fn, strsplit, ".venn.homer.txt"))


#### -- extract total number peaks after merge ------------------------------------------------------------------
path <- "/Volumes/JD_WD2/JRD_CBDMLab/1_JRD_CBDM_PROJECTS/ProjectFolders/T-TregEpigenome/Data/1_Experimental/ATAC/Analyses/A8-all-samples/peaks/replicates-merged-100bp/"

# store filenames with extensions
fn <- list.files(path = path, pattern = ".txt", full.names=T)

# also read into list
fl <- list()
fl <- lapply(fn, read.table, sep="\t")

# extract total number peaks 
merged_peak_totals <- sapply(fl, nrow)


#### -- extract peak numbers from files --------------------------------------------------------------------------
library(gsubfn)
library(dplyr)

# set path
path <- "/Volumes/JD_WD2/JRD_CBDMLab/1_JRD_CBDM_PROJECTS/ProjectFolders/T-TregEpigenome/Data/1_Experimental/ATAC/Analyses/A8-all-samples/peaks/replicates-merged-100bp-venn/"

# store filenames with extensions
fn <- list.files(path = path, pattern = ".txt", full.names=T)

# also read into list
fl <- list()
fl <- lapply(fn, read.table, skip=1, sep="\t")

# store column names
headers <- c("rep2_only", "rep1_only","common","File")

# name columns within list elements
for (i in 1:length(fl)) {
  names(fl[[i]]) <- headers
}

# extract peak numbers from each df in list
totals <- as.data.frame(t(sapply(fl, function(x) unlist(x$common))))

# name columns
names(totals) <- c("rep2_only", "rep1_only","common")

# compute totals
totals$rep1_total <- totals$rep1_only + totals$common
totals$rep2_total <- totals$rep2_only + totals$common
totals$merge_total <- merged_peak_totals
totals$pct_concordant <- (totals$common / totals$merge_total)*100
totals$pct_rep1_only <- (totals$rep1_only / totals$merge_total)*100
totals$pct_rep2_only <- (totals$rep2_only / totals$merge_total)*100

# merge into table
merge_result_table1 <- data.frame(SampleName =  samplename,
                                 MergeDistanct = "100bp",
                                 totals)


#### repeat for 50bp distance
#### -- extract sample information from filenames ----------------------------------------------------------------
# set path
path <- "/Volumes/JD_WD2/JRD_CBDMLab/1_JRD_CBDM_PROJECTS/ProjectFolders/T-TregEpigenome/Data/1_Experimental/ATAC/Analyses/A8-all-samples/peaks/replicates-merged-50bp-venn/"

# read file names
fn <- list.files(path = path, pattern = ".txt")

# extract sample names
samplename <- unlist(lapply(fn, strsplit, ".50bp.venn.homer.txt"))


#### -- extract total number peaks after merge ------------------------------------------------------------------
path <- "/Volumes/JD_WD2/JRD_CBDMLab/1_JRD_CBDM_PROJECTS/ProjectFolders/T-TregEpigenome/Data/1_Experimental/ATAC/Analyses/A8-all-samples/peaks/replicates-merged-50bp/"

# store filenames with extensions
fn <- list.files(path = path, pattern = ".txt", full.names=T)

# also read into list
fl <- list()
fl <- lapply(fn, read.table, sep="\t")

# extract total number peaks 
merged_peak_totals <- sapply(fl, nrow)


#### -- extract peak numbers from files --------------------------------------------------------------------------
library(gsubfn)
library(dplyr)

# set path
path <- "/Volumes/JD_WD2/JRD_CBDMLab/1_JRD_CBDM_PROJECTS/ProjectFolders/T-TregEpigenome/Data/1_Experimental/ATAC/Analyses/A8-all-samples/peaks/replicates-merged-50bp-venn/"

# store filenames with extensions
fn <- list.files(path = path, pattern = ".txt", full.names=T)

# also read into list
fl <- list()
fl <- lapply(fn, read.table, skip=1, sep="\t")

# store column names
headers <- c("rep2_only", "rep1_only","common","File")

# name columns within list elements
for (i in 1:length(fl)) {
  names(fl[[i]]) <- headers
}

# extract peak numbers from each df in list
totals <- as.data.frame(t(sapply(fl, function(x) unlist(x$common))))

# name columns
names(totals) <- c("rep2_only", "rep1_only","common")

# compute totals
totals$rep1_total <- totals$rep1_only + totals$common
totals$rep2_total <- totals$rep2_only + totals$common
totals$merge_total <- merged_peak_totals
totals$pct_concordant <- (totals$common / totals$merge_total)*100
totals$pct_rep1_only <- (totals$rep1_only / totals$merge_total)*100
totals$pct_rep2_only <- (totals$rep2_only / totals$merge_total)*100

# merge into table
merge_result_table2 <- data.frame(SampleName =  samplename,
                                 MergeDistanct = "50bp",
                                 totals)


#### repeat for 0bp distance
#### -- extract sample information from filenames ----------------------------------------------------------------
# set path
path <- "/Volumes/JD_WD2/JRD_CBDMLab/1_JRD_CBDM_PROJECTS/ProjectFolders/T-TregEpigenome/Data/1_Experimental/ATAC/Analyses/A8-all-samples/peaks/replicates-merged-0bp-venn/"

# read file names
fn <- list.files(path = path, pattern = ".txt")

# extract sample names
samplename <- unlist(lapply(fn, strsplit, ".0bp.venn.homer.txt"))


#### -- extract total number peaks after merge ------------------------------------------------------------------
path <- "/Volumes/JD_WD2/JRD_CBDMLab/1_JRD_CBDM_PROJECTS/ProjectFolders/T-TregEpigenome/Data/1_Experimental/ATAC/Analyses/A8-all-samples/peaks/replicates-merged-0bp/"

# store filenames with extensions
fn <- list.files(path = path, pattern = ".txt", full.names=T)

# also read into list
fl <- list()
fl <- lapply(fn, read.table, sep="\t")

# extract total number peaks 
merged_peak_totals <- sapply(fl, nrow)


#### -- extract peak numbers from files --------------------------------------------------------------------------
library(gsubfn)
library(dplyr)

# set path
path <- "/Volumes/JD_WD2/JRD_CBDMLab/1_JRD_CBDM_PROJECTS/ProjectFolders/T-TregEpigenome/Data/1_Experimental/ATAC/Analyses/A8-all-samples/peaks/replicates-merged-0bp-venn/"

# store filenames with extensions
fn <- list.files(path = path, pattern = ".txt", full.names=T)

# also read into list
fl <- list()
fl <- lapply(fn, read.table, skip=1, sep="\t")

# store column names
headers <- c("rep2_only", "rep1_only","common","File")

# name columns within list elements
for (i in 1:length(fl)) {
  names(fl[[i]]) <- headers
}

# extract peak numbers from each df in list
totals <- as.data.frame(t(sapply(fl, function(x) unlist(x$common))))

# name columns
names(totals) <- c("rep2_only", "rep1_only","common")

# compute totals
totals$rep1_total <- totals$rep1_only + totals$common
totals$rep2_total <- totals$rep2_only + totals$common
totals$merge_total <- merged_peak_totals
totals$pct_concordant <- (totals$common / totals$merge_total)*100
totals$pct_rep1_only <- (totals$rep1_only / totals$merge_total)*100
totals$pct_rep2_only <- (totals$rep2_only / totals$merge_total)*100

# merge into table
merge_result_table3 <- data.frame(SampleName =  samplename,
                                 MergeDistanct = "0bp",
                                 totals)

#### -- merge all tables ---------------------------------------------------------------------------------------
mergecompare <- rbind(merge_result_table1, merge_result_table2, merge_result_table3)


#### -- export ---------------------------------------------------------------------------------------------------
path <- "/Volumes/JD_WD2/JRD_CBDMLab/1_JRD_CBDM_PROJECTS/ProjectFolders/T-TregEpigenome/Data/1_Experimental/ATAC/Analyses/A8-all-samples/peaks/"
save(mergecompare, file = paste0(path,"rep_merge_compare_table",".RDa"))

# csv
write.csv(mergecompare, file = paste0(path,"rep_merge_compare_table",".csv"))
```
<br>


#### Plot stats from replicate merge
```{r merge_rep_plots}
library(reshape2)
library(RColorBrewer)
library(scales)
library(ggplot2)
#### -- load summary stat files from merge ------------------------------------------------------------------------
path <- "/Volumes/JD_WD2/JRD_CBDMLab/1_JRD_CBDM_PROJECTS/ProjectFolders/T-TregEpigenome/Data/1_Experimental/ATAC/Analyses/A8-all-samples/peaks/"
load(paste0(path,"rep_merge_compare_table",".RDa"))


#### -- plot stacked bar graph -----------------------------------------------------------------------------------
# copy df
dat <- mergecompare

# melt
datm <- melt(dat, id.vars = c(names(dat[1:8])))

# plot
g <- ggplot(datm, aes(x=SampleName, y= value, fill=variable))
g +
  geom_bar(stat="identity", color="black") + 
  guides(fill=guide_legend(reverse=T)) +
  scale_fill_brewer(palette="Spectral") +
  coord_flip() +
  scale_x_discrete(limits=rev(levels(datm$SampleName))) + 
  ylab("Percent of Total") +
  geom_hline(yintercept=50, color = "black", linetype = 2) +
  facet_wrap(~MergeDistanct) +
  theme_bw() 

path <- "/Volumes/JD_WD2/JRD_CBDMLab/1_JRD_CBDM_PROJECTS/ProjectFolders/T-TregEpigenome/Data/1_Experimental/ATAC/Analyses/A8-all-samples/plots/"
ggsave(paste0(path,"rep_merge_compare_plot",".pdf"), width=10, height=7)

```
<br>


#### Filter to only keep peaks called in both replicates
+ in homer merge output, information as to whether peak was unique or common is kept
+ filter for common
```{r filter_common_peaks}
#### -- import merged peak files into R --------------------------------------------------------------------------
path <- "/Volumes/JD_WD2/JRD_CBDMLab/1_JRD_CBDM_PROJECTS/ProjectFolders/T-TregEpigenome/Data/1_Experimental/ATAC/Analyses/A8-all-samples/peaks/replicates-merged-100bp/"

# store filenames with extensions
fn <- list.files(path = path, pattern = ".txt", full.names=T)
fn1 <- list.files(path = path, pattern = ".txt")


#### -- filter for common peaks ----------------------------------------------------------------------------------

# read into list
fl <- list()
fl <- lapply(fn, read.table, sep="\t")

# set column names
#names(dat) <- headers
headers1 <- read.table(fn[[1]], nrow=1, comment.char="", sep="\t")
headers <- as.character(unlist(headers1))

for (i in 1:length(fl)) {
  names(fl[[i]]) <- headers
}

# filter each df in the list to keep common peaks, by keeping rows that contain "|" pipe in `Parent files`
rep_consistent_peaks <- lapply(fl, function(x) x[grep("\\|", x$`Parent files`),])

# name the list elements to match the files
names(rep_consistent_peaks) <- gsub("\\.homer.txt","", fn1)


#### -- save list of peak files  ----------------------------------------------------------------------------------
# as RDa
path <- "/Volumes/JD_WD2/JRD_CBDMLab/1_JRD_CBDM_PROJECTS/ProjectFolders/T-TregEpigenome/Data/1_Experimental/ATAC/Analyses/A8-all-samples/peaks/final-replicate-merge/"
save(rep_consistent_peaks, file = paste0(path, "reps_consistent_peaks_list",".RDa"))


#### -- export as separate tdt files for each sample --------------------------------------------------------------
path <- "/Volumes/JD_WD2/JRD_CBDMLab/1_JRD_CBDM_PROJECTS/ProjectFolders/T-TregEpigenome/Data/1_Experimental/ATAC/Analyses/A8-all-samples/peaks/final-replicate-merge/"

# convert each df to bed format, [p]eak[l]ist_bed
pl_bed <- lapply(rep_consistent_peaks, function(x) {
  x <- x[,c(2,3,4,1,5)]
})

lapply(names(pl_bed),
       function(x) write.table(pl_bed[x], file=paste(paste0(path,x),"bed",sep="."), sep = "\t", 
                              quote=F, row.names=F,col.names=c("#chr","start","end","name","strand")))
```
<br>


#### Get read counts for merged peak sets from bam files
+ ran ChIPQC on Orchestra module `stats/R/3.2.1`, out of folder `/groups/cbdm-db/jrd26/ATAC_crossbatch/temp`
+ used separate script that will loop over pairs of peak/bam files: `jd_chipqc_peaks.R`
+ for script to work, had to duplicate merged peak files and give each unique name, so appended "1" or "2"
+ output = .csv file for each replicate's read counts in its merged peak set
+ command:
+ `bsub -q priority -W 10:00 -R "rusage[mem=64000]" Rscript jd_chipqc_peaks.R`
<br> 


#### Generate tables of replicate read counts in merged peak sets
```{r plot_reps_after_merge}
#### -- import files ---------------------------------------------------------------------------------------------
path <- "/Volumes/JD_WD2/JRD_CBDMLab/1_JRD_CBDM_PROJECTS/ProjectFolders/T-TregEpigenome/Data/1_Experimental/ATAC/Analyses/A8-all-samples/peaks/final-replicate-merge/reps-merged-with-counts/rep1/"

# generate list of filenames
fl <- list.files(path = path, pattern = ".csv", full.names = T)

#### --read_peak_csv function ------------------------------------------------------------------------------------

# function to take in file, extract sample info, and build table with read counts for each peak
# ARGS:
# file = filename, with full path name and extension
read_peak_csv <- function(file){
  # read in filename
  fn <- unlist(strsplit(file, split="*//"))[2]
  
  # read in file
  dat <- read.csv(file = file)
  
  # extract sample name info from filename
  samplename1 <- unlist(strsplit(fn, "*.csv"))[[1]]
  # remove number at end of name
  samplename <- unlist(strsplit(samplename1, split = "[0-9]$"))
  
  # expand name vector so that length matches # of peaks in file
  samplename <- rep(samplename, nrow(dat))
  
  # generate peaknames 
  peaknames <- paste(samplename, "_peak_",1:nrow(dat),sep="")
  
  # make new df with peaks and reads
  peak_df <- data.frame(Sample = samplename,
                        dat[,c("seqnames","start","end")],
                        rep1rpm = dat[,"rpm"],
                        peakID = peaknames)
}


#### -- apply over file list ------------------------------------------------------------------------------------
pl1 <- lapply(fl, read_peak_csv) #[p]eak[l]ist[1]


#### -- get rpm for rep2 ----------------------------------------------------------------------------------------
path <- "/Volumes/JD_WD2/JRD_CBDMLab/1_JRD_CBDM_PROJECTS/ProjectFolders/T-TregEpigenome/Data/1_Experimental/ATAC/Analyses/A8-all-samples/peaks/final-replicate-merge/reps-merged-with-counts/rep2/"

# generate list of filenames
fl <- list.files(path = path, pattern = ".csv", full.names = T)

# read files into list of dfs
datlist <- lapply(fl, read.csv)

# loop over and extract rpm
rpm2_list <- lapply(datlist, function(x) x$rpm)

# add rep2 rpm to peak table
pl2 <- list() #[p]eak[l]ist[2]
for(i in 1:length(pl1)) {
  pl2[[i]] <- cbind(pl1[[i]], rep2rpm = rpm2_list[[i]])
}

# change order of columns in peak table
pl3 <- list()
for(i in 1:length(pl2)) {
  pl3[[i]] <- data.frame(pl2[[i]][,c(1:4,6,5,7)])
}

# name elements of list
# generate list of filenames
fl <- list.files(path = path, pattern = ".csv", full.names = T)
fn <- lapply(fl, function(x) unlist(strsplit(x, split="*//"))[2])
fn <- unlist(lapply(fn, function(x) unlist(strsplit(x, split="*.csv"))[1]))

names(pl3) <- fn 

#### -- export list of peak tables -------------------------------------------------------------------------------
path <- "/Volumes/JD_WD2/JRD_CBDMLab/1_JRD_CBDM_PROJECTS/ProjectFolders/T-TregEpigenome/Data/1_Experimental/ATAC/Analyses/A8-all-samples/peaks/final-replicate-merge/reps-merged-with-counts/"
file <- "reps-merged-peaklist"
save(pl3, file = paste0(path,file,".RDa"))

```
<br>


#### Sanity check on genome browser
+ for each sample (cell type):
  + compare peaks called in both reps (from merged file) vs peaks called in only one 
  + sort peaks by rpm, validate strong and weak peaks
  + if replicate read counts differ, validate that ChIPQC quantitation matches browser scale
+ look at some peaks across multiple samples (e.g. strongest mac in the tregs and vice versa)
+ used IGV and .bigwig files 
+ so far, colonNrp1n, mus, 
+ anecdotally, are peaks only called in one are convincing in both reps
<br>


#### Plot replicate pair-wise correlations of read counts in merged peak sets
```{r plot_reps_corr}
library(reshape2)
library(scales)
library(RColorBrewer)
library(ggplot2)

#### -- import list of peak tables -------------------------------------------------------------------------------
path <- "/Volumes/JD_WD2/JRD_CBDMLab/1_JRD_CBDM_PROJECTS/ProjectFolders/T-TregEpigenome/Data/1_Experimental/ATAC/Analyses/A8-all-samples/peaks/final-replicate-merge/reps-merged-with-counts/"
file <- "reps-merged-peaklist"
load(file = paste0(path,file,".RDa"))

#### -- get into shape for plotting -------------------------------------------------------------------------------
plm <- melt(pl3, id.vars=names(pl3[[1]]))[-8] #[p]eak[l]ist[m]elted

# calculate vector of r2 values
reps_pearson <- unlist(lapply(pl3, function(x) cor(x$rep1rpm,x$rep2rpm)))

# save
path <- "/Volumes/JD_WD2/JRD_CBDMLab/1_JRD_CBDM_PROJECTS/ProjectFolders/T-TregEpigenome/Data/1_Experimental/ATAC/Analyses/A8-all-samples/plots/"
png(filename = paste0(path,"rep_scatter_with_cor",".png"), width=1000, height=1000)

g <- ggplot(plm, aes(x= log(rep1rpm,2), y= log(rep2rpm,2), color = seqnames))
g + 
  geom_point(size=0.6, alpha=0.6) + 
  facet_wrap(~Sample) + 
  xlab("Rep1 log2 rpm") +
  ylab("Rep2 log2 rpm") + 
  #annotate with cor
  #add pretty breaks
  #change color to fill?
  theme_bw()
dev.off()


```
<br>


<br>
```{r test}
path <- "/Volumes/JD_WD2/JRD_CBDMLab/1_JRD_CBDM_PROJECTS/ProjectFolders/T-TregEpigenome/Data/1_Experimental/ATAC/Analyses/A8-all-samples/testing/"
peaks <- list.files(path=path,pattern = "*.bed.txt")














  