---
title: "ATAC peaks"
author: "JRD"
date: "February 4, 2016"
output: html_document
---

last modified: `r Sys.Date()`
<br>    

#### TO DO
+ If chipqc script works, want to be able to have peaks and reads folders read in from command line
<br>    

#### R Markdown Specifications
Set global options
```{r setoptions, echo=F}
library(knitr)
opts_chunk$set(echo=F, eval=F, warning=F, message=F)
```
<br>    


Github
https://github.com/jdis24/ATAC.git
+ can't figure out how to link Rstudio project to github subdirectory, rather than home dir
<br>    

#### Script Overview: Step (tool)
+ name = `atac_peaks_PE.sh`
+ make directory of reads by chromosome (homer)
+ make bedGraph file (homer)
+ make bigWig file (UCSC tools)
+ find peaks in DNAse mode (homer)
+ find peaks in factor mode (homer)
+ merge results of dnase and factor modes (homer)
+ compress output files
+ Orchestra: 
`bsub -q priority -W 1:00 ./atac_peaks_PE.sh .`
<br>    


#### Script
+ default = hidden
```{r atac_peaks_PE.sh, engine="bash"}
#!/bin/bash
# run where the .sorted.uniq.bam.gz files are
# usage: ./atac_peaks_PE.sh  dir_where_files_are prefix
# 
#
# author: Joanna R. DiSpirito
# last modified: 2/4/16


#### Load modules
source ~/.bash_profile

module load seq/samtools/0.1.19
module load seq/BEDtools/2.19.0
module load seq/homer/4.6
module load seq/UCSC-tools

#### Set working directory and prefix for output files (args)
echo "working directory : " $1
echo "prefix : " $2
prefix=$2

cd $1

#### Gunzip files
echo "Decompressing files..."
gunzip -fv *sorted.uniq.bam.gz


#### makeTagDirectory and generate QC files
echo "Making Tag Directory..."
makeTagDirectory $prefix.homer.tagdir/ -genome /groups/shared_databases/genomes/mm9.fa -tbp 1 -checkGC -illuminaPE *sorted.uniq.bam &> $prefix.tagdir.stdout.txt


#### make bedGraph file
echo "Making bedGraph file..."
makeUCSCfile $prefix.homer.tagdir/ -fsize 1e50 -norm 1e6 -name $prefix -o $prefix.bedGraph


#### make bigWig file
echo "Making biwig file..."

# move chrom sizes from jrd/resources directory to current directory
cp /groups/cbdm_lab/jrd26/JRD_Resources/mm9.chrom.sizes .

# unzip bedGraph
gunzip -fv $prefix.bedGraph.gz

# sort bedGraph and delete last line
sort -k1,1 -k2,2n $prefix.bedGraph >  $prefix_sorted1.bedGraph
sed '$d'  $prefix_sorted1.bedGraph  > $prefix_sorted.bedGraph

# convert to bw
bedGraphToBigWig $prefix_sorted.bedGraph mm9.chrom.sizes $prefix.bw


#### Find peaks (altered original script to add different settings)
echo "Finding peaks in factor mode..."
findPeaks $prefix.homer.tagdir/ -style factor -norm 1000000 -o $prefix.peaks.factor.txt &> $prefix.stdout.peaks.factor.txt

echo "Finding peaks in histone mode..."
findPeaks $prefix.homer.tagdir/ -style histone -norm 1000000 -o $prefix.peaks.histone.txt &> $prefix.stdout.peaks.histone.txt

echo "Finding peaks in dnase mode..."
findPeaks $prefix.homer.tagdir/ -style dnase -norm 1000000 -o $prefix.peaks.dnase.txt &> $prefix.stdout.peaks.dnase.txt

echo "Merging results of dnase and factor modes..."
mergePeaks $prefix.peaks.dnase.txt $prefix.peaks.factor.txt -o $prefix.peaks.atac.txt &> $prefix.stdout.merge.txt
 

#### Compress files
echo "zipping bedGraph file..."
gzip $prefix.bedGraph

echo "zipping bigWig file..."
gzip $prefix.bw

echo "zipping bam file..."
gzip *sorted.uniq.bam


########################### DEPRECATED ######################################################
# deprecated Amit lab settings
# findPeaks homer_tagdir/ -L 0 -C 3 -size 1000 -minDist 1000 -tbp 1 -o peaks.txt

# deprecated DPZ settings
# echo "Finding peaks using user settings..."
# findPeaks $prefix.homer.tagdir/ -C 3 -norm 1000000 -o $prefix.peaks.user.txt &> $prefix.stdout.peaks.user.txt
```
<br>    


#### getting sample-by-sample normalized read counts for each peak
+ checked output of one sample on genome browser
+ looks good - calculated rpm values correspond to peak height
+ currently running in orchestra
+ `bsub -q mcore -W 4:00 -n 8 Rscript jd_chipqc_peaks.R`
+ was in PEND for hour so tried
+ `bsub -q priority -W 10:00 -R "rusage[mem=64000]" Rscript jd_chipqc_peaks.R` 
+ this works, currently running as separate script in Orchestra, code pasted below
```{r jd_chipqc_peaks.R}
# load libraries
library(ChIPQC)

# function to generate chipqc object and peak table with read counts in rpm 
# exports ChIPQCsample object and peak table
# ARGS: 

peak_rpm <- function(reads, peaks) {
  # create chromosome vector
  chr <- paste("chr",1:19,sep="")
  chr <- c(chr, "chrX","chrY")
  
  # create chipQCsample object
  qcs <- ChIPQCsample(reads, peaks, annotation = "mm9", chromosomes=chr)
  
  # extract total reads
  tr <- QCmetrics(qcs)[[1]]
  
  # extract peak table with counts
  qcs_peaks <- peaks(qcs)
  
  # convert to dataframe
  qcs_peaks <- as.data.frame(qcs_peaks)
  
  # append total reads
  qcs_peaks$Total_Rds <- tr
  
  # calculate and append rpm
  qcs_peaks$rpm <- (qcs_peaks$Counts / qcs_peaks$Total_Rds) * 1000000
    
  # extract df name 
  Sample <- strsplit(peaks, split=".peaks.atac.bed.txt")
  Sample <- unlist(Sample)
  
  # append name
  qcs_peaks <- cbind(Sample, qcs_peaks)
  
  # write out csv file
  write.csv(qcs_peaks, file=paste(Sample,".csv", sep=""), row.names=F)
  
  # write ChIPQCsample
  save(qcs, file=paste(Sample,".RDa",sep=""))
}

#### Function implementation ####
# set working directory
setwd("/groups/cbdm-db/jrd26/ATAC_crossbatch/R/temp/")

# get names of peak files
peaks <- list.files(".", pattern = "*.bed.txt")

# get names of bam files
reads <- list.files(".", pattern = "*.uniq.bam$")

# loop over each peak/bam pair in folder
for(i in 1:length(reads)) {
  peak_rpm(reads[i], peaks[i])
}
```
<br>    

#### Import peak files into peak list
```{r peak_list, eval=T}
# create path to .csv files  
path <- "/Volumes/JD_WD2/JRD_CBDMLab/1_JRD_CBDM_PROJECTS/ProjectFolders/T-TregEpigenome/Data/1_Experimental/ATAC/Master_peaks/R/temp/"

# get filenames
filenames1 <- list.files(path = path, pattern = ".csv")
filenames <- paste(path, filenames, sep="") # update filenames with path

# combine peak table .csv dfs into a list
colClasses_peaks <- c("factor","character", "factor", rep("numeric",2), "factor", rep("numeric",2)) # set column classes
pl <- lapply(filenames, read.csv, colClasses = colClass) # [p]eak [l]ist

# name the list elements to match the files
names(pl) <- gsub("\\.csv", "", filenames1)
```
<br>    


#### Generate table of peak numbers for range of rpm threshold values
+ goal = readable table exported as .csv
+ goal = readable table printed in .html output of knitr
```{r threshold_table, eval=T, results="asis"}
library(xtable)

# function to compute the number of peaks >=n rpm in each sample
# ARGS: df = dataframe of peak height rpms, peak_height_var = string of var name storing rpm values, cutoffs = numeric vector of rpm cutoffs 
# USAGE: number_peaks_above_nrpm(df, "peak_height_var", cutoffs)
number_peaks_above_nrpm <- function(df, rpm_var, cutoffs) { 
  # create indexing vector representing number of elements in cutoffs
  n <- length(cutoffs) 
  # create output vector whose length = length of cutoffs
  o <- numeric(n) 
  # create variable to store vector of extracted rpm values
  rpm <- df[[rpm_var]]
  # loop over each element of cutoffs and count number rows where peak height >= that element
  for(i in 1:n) { 
    o[i] <- as.numeric(sum(rpm >= cutoffs[[i]])) # calculate the total number of rows where rpm >= ith value in cutoffs  
  }
  # print output vector
  return(o) 
}


# create cutoff vector of rpm values
cutoffs <- c(seq(0,5,by=0.5))

# apply cutoffs to list
peaks_pass_filter <- sapply(pl, number_peaks_above_nrpm, "rpm", cutoffs)

# transform list into dataframe (if doesn't work may have to call base as.data.frame bc of Biocgenerics masking?)
ppf <- as.data.frame(peaks_pass_filter) #[p]eaks [p]ass [f]ilter
ppf <- cbind(ppf,"rpm_cutoff" = cutoffs) # add column for each cutoff

# write to file
write.csv(ppf, "peak_totals_rpm_filtered.csv")

# generate xtable
print(xtable(ppf), type="html", include.rownames=F)
```
<br>    


#### Plot distribution of read counts for each sample
+ goal = boxplots
+ goal = histograms, bin size and limits will be key for clear data
```{r plots, eval=T}
library(reshape2)
library(ggplot2)
library(dplyr)

# get peak list into shape for plotting
plm <- melt(pl, id.vars=names(pl[[1]])) # [p]eak [l]ist [m]elted
plm$L1 <- NULL # remove extraneous column

# calculcate log2 rpm values and append as new column



```
<br>    


#### Generate ChIPQC summary table 
+ TO DO:






