---
title: "ATAC_pairwise"
author: "JRD"
date: "February 17, 2016"
output: html_document
---

last modified: `r Sys.Date()`
<br>   

#### README
+ this document contains code and explanation for the pair-wise analysis of each tissue Treg ATAC dataset with its corresponding splenic Treg dataset
+ for details on how the input files used in this analysis were generated, see *ATAC_mapping.Rmd* and *ATAC_peaks.Rmd*
<br>   

#### R Markdown Specifications
Set global options
```{r setoptions, echo=F}
library(knitr)
opts_chunk$set(echo=F, eval=F, warning=F, message=F)
```
<br>    


#### Pseudo
**call peaks and get read counts**
+ [these steps were done in "ATAC_peaks.Rmd"]
+ for each sample, call peaks in each replicate
+ then merge into single peakset
+ for each replicate, count reads in each peak, normalized to total library size
+ export .csv files of peaks and counts    

**filter for strong peaks, merge peaks and call differentially accessible peaks**
+ [all subsequent steps done in this document]
+ plot distributions of read counts 
+ filter for strong peaks (start with >=2.5 rpm)
+ get genomic annotations of strong peaks (compare to unfilterd to check degree of TSS bias in filtered set)
+ use DiffBind to call differentially accessible peaks
<br>    

#### Special functions for plotting
```{r plotting_functions}
# plot multiple plots per page
# ggplot objects can be passed in ..., or to plotlist (as a list of ggplot objects)
# - cols:   Number of columns in layout
# - layout: A matrix specifying the layout. If present, 'cols' is ignored.
#
# If the layout is something like matrix(c(1,2,3,3), nrow=2, byrow=TRUE),
# then plot 1 will go in the upper left, 2 will go in the upper right, and
# 3 will go all the way across the bottom.
multiplot <- function(..., plotlist=NULL, file, cols=1, layout=NULL) {
  library(grid)

  # Make a list from the ... arguments and plotlist
  plots <- c(list(...), plotlist)

  numPlots = length(plots)

  # If layout is NULL, then use 'cols' to determine layout
  if (is.null(layout)) {
    # Make the panel
    # ncol: Number of columns of plots
    # nrow: Number of rows needed, calculated from # of cols
    layout <- matrix(seq(1, cols * ceiling(numPlots/cols)),
                    ncol = cols, nrow = ceiling(numPlots/cols))
  }

 if (numPlots==1) {
    print(plots[[1]])

  } else {
    # Set up the page
    grid.newpage()
    pushViewport(viewport(layout = grid.layout(nrow(layout), ncol(layout))))

    # Make each plot, in the correct location
    for (i in 1:numPlots) {
      # Get the i,j matrix positions of the regions that contain this subplot
      matchidx <- as.data.frame(which(layout == i, arr.ind = TRUE))

      print(plots[[i]], vp = viewport(layout.pos.row = matchidx$row,
                                      layout.pos.col = matchidx$col))
    }
  }
}

# volcano plots from ggplot2
# ARGS: cr = [c]ontrast[r]esults from Diffbind, as data frame
#       n = fold-change cutoff
volcano <- function(cr,n) {
  FCa <- as.numeric(sum(cr$Fold >=n ))    # calc number peaks >=n ; [F]old[C]hange[a]/b
  FCb <- as.numeric(sum(cr$Fold <= (-n))) # number <=2 ; [F]old[C]hange[b]/a
  
  g <- ggplot(cr, aes(x=Fold, y=log(p.value,10),  # add data to plot
                      color=c(Fold>=n | Fold <=(-n))))
  
  g + geom_point(alpha=0.5) +   # plot
  theme_bw() +
  scale_color_manual(values=c('black','red')) + 
  xlab(paste(Tissue[1],'/',Tissue[4],' (log2 FC)',sep="")) + 
  ylab('log10 pvalue') +
  scale_x_continuous(limits=c(-8,8),breaks=seq(-8,8,by=1)) + 
  annotate("text", x=6, y=0, label=FCa, color="red") +
  annotate("text", x=-6, y=0, label=FCb, color="red") +
  guides(color=FALSE)
}

```
<br>  




#### Import peak files into peak list and export bed files
```{r peak_list, eval=T}
# create path to .csv files  
path <- "/Volumes/JD_WD2/JRD_CBDMLab/1_JRD_CBDM_PROJECTS/ProjectFolders/T-TregEpigenome/Data/1_Experimental/ATAC/Master_peaks/R/peaks_with_counts/working/"

# get filenames
filenames1 <- list.files(path = path, pattern = ".csv")
filenames <- paste(path, filenames1, sep="") # update filenames with path

# combine peak table .csv dfs into a list
colClass <- c(rep("factor",2), rep("numeric",3), "character", rep("numeric",4)) # set column classes
pl <- lapply(filenames, read.csv, colClasses = colClass) # [p]eak [l]ist

# name the list elements to match the files
names(pl) <- gsub("\\.csv", "", filenames1)

# convert each df to bed format
pl_bed <- lapply(pl, function(x) {
  x <- x[,c("seqnames","start","end")]
})

# export each df as tdt bed file
path <- "/Volumes/JD_WD2/JRD_CBDMLab/1_JRD_CBDM_PROJECTS/ProjectFolders/T-TregEpigenome/Data/1_Experimental/ATAC/Analyses/A6-vat-spleen/bed/"

lapply(names(pl_bed),
       function(x) write.table(pl_bed[x], file=paste(paste0(path,x),"bed",sep="."), sep = "\t", 
                              quote=F, row.names=F,col.names=c("#chr","start","end")))
```
<br>    


#### Generate table of peak numbers for range of rpm threshold values
```{r threshold_table, eval=T, results="asis"}
library(xtable)

# function to compute the number of peaks >=n rpm in each sample
# ARGS: df = dataframe of peak height rpms, peak_height_var = string of var name storing rpm values, cutoffs = numeric vector of rpm cutoffs 
# USAGE: number_peaks_above_nrpm(df, "peak_height_var", cutoffs)
number_peaks_above_nrpm <- function(df, rpm_var, cutoffs) { 
  # create indexing vector representing number of elements in cutoffs
  n <- length(cutoffs) 
  # create output vector whose length = length of cutoffs
  o <- numeric(n) 
  # create variable to store vector of extracted rpm values
  rpm <- df[[rpm_var]]
  # loop over each element of cutoffs and count number rows where peak height >= that element
  for(i in 1:n) { 
    o[i] <- as.numeric(sum(rpm >= cutoffs[[i]])) # calculate the total number of rows where rpm >= ith value in cutoffs  
  }
  # print output vector
  return(o) 
}


# create cutoff vector of rpm values
cutoffs <- c(seq(0,5,by=0.5))

# apply cutoffs to list
peaks_pass_filter <- sapply(pl, number_peaks_above_nrpm, "rpm", cutoffs)

# transform list into dataframe (if doesn't work may have to call base as.data.frame bc of Biocgenerics masking?)
ppf <- as.data.frame(peaks_pass_filter) #[p]eaks [p]ass [f]ilter
ppf <- cbind(ppf,"rpm_cutoff" = cutoffs) # add column for each cutoff

# write to file
path <- "/Volumes/JD_WD2/JRD_CBDMLab/1_JRD_CBDM_PROJECTS/ProjectFolders/T-TregEpigenome/Data/1_Experimental/ATAC/Analyses/A6-vat-spleen/"
# set sample names for file export
sample_name <- unique(unlist((strsplit(filenames1, split = '_Treg_rep[1,2].csv'))))
sample_name <- unlist(strsplit(sample_name, split='_w???'))

write.csv(ppf, file = paste0(path,sample_name[1],'-vs-',sample_name[3],'_peak_totals_rpm_filtered.csv',sep=""), row.names=F)

# generate xtable
print(xtable(ppf), type="html", include.rownames=F)

# plot distributions of total peaks called at each cutoff
library(reshape2)
library(ggplot2)
ppfm <- melt(ppf, id.vars='rpm_cutoff',variable.name='no_peaks') # ppf[m]elted 
g <- ggplot(ppfm, aes(x=as.factor(rpm_cutoff),y=value))
g + geom_boxplot() +
  ylab('Peaks passing filter') +
  xlab('rpm cutoff') + 
  theme_bw()
ggsave(paste0(path,sample_name[1],'-vs-',sample_name[3], ".peakcut.boxplot.pdf"), width=7, height=5) # export pdf
```
<br>     


#### Plot distribution of read counts for each sample
```{r plots, eval=T}
library(reshape2)
library(ggplot2)
library(dplyr)
# set path for plots 
path <- "/Volumes/JD_WD2/JRD_CBDMLab/1_JRD_CBDM_PROJECTS/ProjectFolders/T-TregEpigenome/Data/1_Experimental/ATAC/Analyses/A6-vat-spleen/"

# get peak list into shape for plotting
plm <- melt(pl, id.vars=names(pl[[1]])) # [p]eak [l]ist [m]elted
plm$L1 <- NULL # remove extraneous column

# calculcate log2 rpm values and append as new column
plm$log2_rpm <- log(plm$rpm, 2)

# show number of peaks where rpm =0 
rpm0pks <- plm$log2_rpm == "-Inf"
plm_rpm0pks <- plm[rpm0pks,]

print("Peaks where rpm =0 ...") 
xtable(plm_rpm0pks)

# filter out peaks where rpm = 0
plmf <- plm[!rpm0pks,]  #plm[f]iltered

# draw boxplot of all peaks in each sample
g <- ggplot(data = plmf, aes(x=Sample, y=log2_rpm, fill=Sample))
g + geom_boxplot(varwidth =T) +
  theme_bw() + 
  ylim(-2,10) +
  theme(axis.title.x = element_blank())
ggsave(paste0(path,sample_name[1],'-vs-',sample_name[3], ".allpeaks.boxplot.pdf"), width=7, height=5) # export pdf

# draw boxplot of peaks passing n rpm
# define function to filter peaks and draw plot
bp_nrpm <- function(df, n) {
  df %>%
    filter(rpm >= n) %>% # filter by rpm cutoff 
    ggplot(aes(x=Sample, y=log2_rpm, fill=Sample)) + # plot
    geom_boxplot(varwidth=T) +
    ylim(-2,10) +
    theme_bw() + 
    theme(axis.title.x = element_blank())
}

n <- 2.5 # set rpm cutoff
bp_nrpm(plmf, n) # draw plot
ggsave(paste0(path,sample_name[1],'-vs-',sample_name[3],'.',n,"rpm.filtered.pdf"), width =7, height =5) # export pdf
# ggsave(paste0(path, "a4-boxplot-rpm-filtered.pdf"), width=5, height=4)


# draw histograms of all peaks in each sample
g <- ggplot(data = plmf, aes(x=log2_rpm))
g + geom_histogram(fill="white",color="black",binwidth=0.1) +
  facet_grid(Sample~.) +
  geom_vline(xintercept =log(1,2)) +
  geom_vline(xintercept = log(2,2)) +
  geom_vline(xintercept = log(3,2)) +
  theme_bw()
ggsave(paste0(path,sample_name[1],'-vs-', sample_name[3], ".allpeaks.hist.pdf"), width=12, height=10)
```
<br>    


#### Filter for strong peaks
```{r filter_peaks, eval=T}
library(dplyr)
# select peaks >= 2.5rpm
pl2p5 <- lapply(pl, filter, rpm >=2.5)

# export each df as .csv file
path <- "/Volumes/JD_WD2/JRD_CBDMLab/1_JRD_CBDM_PROJECTS/ProjectFolders/T-TregEpigenome/Data/1_Experimental/ATAC/Analyses/A6-vat-spleen/csv"
lapply(names(pl2p5),
       function(x) write.csv(pl2p5[[x]], file=paste(paste0(path,x),"2p5rpm","csv",sep="."), row.names=F))

# convert each df to bed format
pl2p5bed <- lapply(pl2p5, function(x) {
  x <- x[,c("seqnames","start","end")]
})

# export each df as tdt bed file
path <- "/Volumes/JD_WD2/JRD_CBDMLab/1_JRD_CBDM_PROJECTS/ProjectFolders/T-TregEpigenome/Data/1_Experimental/ATAC/Analyses/A6-vat-spleen/bed/rpm-filtered/"
lapply(names(pl2p5bed ),
       function(x) write.table(pl2p5bed[x], file=paste(paste0(path,x),"2p5rpm","bed",sep="."), sep = "\t", 
                              quote=F, row.names=F,col.names=c("#chr","start","end")))
```
<br>    


#### Get genomic annotations of all peaks and strong peaks
+ are they biased for TSS?
+ HOMER annotatePeaks.pl
+ will write a script to loop over many bed files
+ for now ran each one-by-one
```{r annotate_pks, engine="bash"}
annotatePeaks.pl <filenmame> mm9 -annStats <filenmame>.annStats.txt > <filenmame>.annpks.txt

annotatePeaks.pl <filenmame>.bed mm9 -annStats <filenmame>.annStats.txt > <filenmame>.annpks.txt
```
<br> 


#### Create merged peak set and perform differential accessibility analysis
+ DiffBind 
+ script called `db.pairwise.R`
```{r merge_and_count}
library(DiffBind)

# create sample sheet
path <- "/Volumes/JD_WD2/JRD_CBDMLab/1_JRD_CBDM_PROJECTS/ProjectFolders/T-TregEpigenome/Data/1_Experimental/ATAC/Analyses/A6-vat-spleen/"

Peaks <- list.files(paste0(path, "bed/rpm-filtered/"))

# create columns for sample sheet
SampleID <- unlist(strsplit(Peaks, split = ".2p5rpm.bed"))
Tissue <- c(rep('Spleen',2),rep('VAT',2))
Factor <- unlist(strsplit(SampleID, split = "_rep1"))
Factor <- unlist(strsplit(Factor, split = "_rep2"))
Condition <- "Treg"
Treatment <- c(rep('Batch1',4))
Replicate <- rep(c('1','2'),2)
PeakCaller <- "homer"
PeakFormat <- "bed"
PeakPath <- list.files(paste0(path, "bed/rpm-filtered"), full.names = T)
bamPath <- list.files(paste0(path,"reads"), full.names=T)

samples <- cbind(SampleID = SampleID, 
                 Tissue = Tissue, 
                 Factor = Factor, 
                 Condition = Condition,
                 Treatment = Treatment,
                 Replicate = Replicate,
                 bamReads = bamPath,
                 Peaks = PeakPath,
                 PeakCaller = PeakCaller,
                 PeakFormat = PeakFormat)
write.csv(samples, file = paste0(path,"samples.csv"), row.names=F)


##-----loading data-----------------------------------------------------------------------------------
# create DBA object
pairw <- dba(sampleSheet = paste0(path,"samples.csv"), bRemoveRandom = T, minOverlap = 2)

# make consensus peakset from replicates
pairw <- dba.peakset(pairw, consensus = -DBA_REPLICATE)

# make consensus peakset from samples and count reads
pairw <- dba.count(pairw, peaks = pairw$masks$Consensus, score = DBA_SCORE_RPKM)

# save dba to file
dba.save(pairw, file=paste(Tissue[1],'-vs-',Tissue[4],sep=""), dir=path, pre="dba_", ext="RData", bMinimize=F)

# export consensus peakset with read counts
pairw_con <- dba.peakset(pairw, pairw$masks$Consensus, bRetrieve = T, 
                       writeFile=paste(path,Tissue[1],'-vs-',Tissue[4],".consensus.rpkm.txt",sep=""), 
                       DataType=DBA_DATA_FRAME)
write.csv(pairw_con, file=paste(path, Tissue[1],'-vs-',Tissue[4], '.consensus.rpkm.csv',sep=""), row.names=F)

# save read count correlation heatmap
pdf(paste(path,Tissue[1],'-vs-',Tissue[4],'.reads.heat.pdf',sep=""), width=10, height=10, pagecentre = T)
par(oma = c(6,6,6,6))
dba.plotHeatmap(pairw)
dev.off()

##------run differential affiniity analysis -----------------------------------------------------------------------
# establish contrast using factor (==Cell Type) metadata
contrast <- dba.contrast(pairw, categories = DBA_FACTOR, minMembers=2)

# perform differential analysis
a1 <- dba.analyze(contrast) # [a]nalysis1

# return contrast results (p-value,FDR,fold) for all peaks
cr <- data.frame(dba.show(a1, bContrasts = T)) # [c]ontrast [r]esults
# write output to file
write.table(cr, file=paste(path,Tissue[1],'-vs-',Tissue[4],'.contrasts.txt',sep=""), sep="\t", row.names=F)

# export contrast results
report <- dba.report(a1, th=1, bUsePval = F, bNormalized = T, bCounts = T,
                     initString= paste(path,Tissue[1],'-vs-',Tissue[4],sep=""),
                     file='contrast.allpeaks')

# export plots from contrast 
# set label
label <- paste(Tissue[1],'-vs-', Tissue[4],sep="")

# set fold-change and fdr cutoffs
fold <- 3
fdr <- 0.01

# XY plot                
dba.plotMA(a1, contrast = 1, fold=fold, bXY=T, th=fdr) 
dev.print(pdf, file=paste(path,label,'_','FC',fold,'.xy.pdf',sep=""), width=5,height=5) 

# MA plot
dba.plotMA(a1, contrast = 1, fold=fold, bXY=F, th=fdr) 
dev.print(pdf, file=paste(path,label,'_','FC',fold,'.ma.pdf',sep=""), width=5,height=5) 

# heatmap of sites passing FC and FDR thresholds
par(oma = c(12,3,4,3))

dba.plotHeatmap(a1, report=dba.report(a1, contrast=1, th = fdr, bUsePval=F, fold=fold), correlations =F,
                bLog=T, score=DBA_SCORE_RPKM, sortFun = var)
dev.print(pdf, file=paste(path,label,'_','FC',fold,'_','FDR',fdr,'.pdf',sep=""))
dev.print(png, file=paste(path,label,'_','FC',fold,'_','FDR',fdr,'.png',sep=""),width=600,height=600)
```
<br>    


#### Plot results of differential accessibility analysis
```{r diffbind_results}
library(ggplot2)

# set path
path <- "/Volumes/JD_WD2/JRD_CBDMLab/1_JRD_CBDM_PROJECTS/ProjectFolders/T-TregEpigenome/Data/1_Experimental/ATAC/Analyses/A6-vat-spleen/"

# import results of dba contrast
filename <- paste(Tissue[1],'-vs-', Tissue[4],'_contrast.allpeaks.csv',sep="")
cr <- read.csv(file=paste0(path,filename),header=T) #[c]ontrast[r]esults

# generate plots for FC=2, 4 and 8
v_fc2 <- volcano(cr, 1) 
v_fc4 <- volcano(cr, 2)
v_fc8 <- volcano(cr ,3)

# plot on single page and print
pdf(paste0(path,sample_name[1],'-vs-', sample_name[3], ".fc.pval.pdf"), width=12, height=10)
multiplot(v_fc2, v_fc4, v_fc8, cols=3)
dev.off()

# print
(paste0(path,sample_name[1],'-vs-', sample_name[3], ".fc.pval.pdf"), width=12, height=10)
```
<br>  


#### Motifs enriched in differentially accessible peaks
+ TODO
<br>  















